{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask\n",
    "import dask.dataframe as dd\n",
    "from scipy import stats\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import shutil\n",
    "import glob\n",
    "import gimmemotifs\n",
    "from pathlib import Path\n",
    "import qnorm\n",
    "from sklearn import datasets, linear_model\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import mean_squared_error, r2_score, average_precision_score\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hue_regplot(data, x, y, hue, palette=None, **kwargs):\n",
    "    from matplotlib.cm import get_cmap\n",
    "    \n",
    "    regplots = []\n",
    "    \n",
    "    levels = data[hue].unique()\n",
    "    \n",
    "    if palette is None:\n",
    "        default_colors = get_cmap('tab10')\n",
    "        palette = {k: default_colors(i) for i, k in enumerate(levels)}\n",
    "    \n",
    "    for key in levels:\n",
    "        regplots.append(\n",
    "            sns.regplot(\n",
    "                x=x,\n",
    "                y=y,\n",
    "                fit_reg=False, \n",
    "                data=data[data[hue] == key],\n",
    "                color=palette[key],\n",
    "                **kwargs\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    return regplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3'\n",
    "Path(f\"{output_dir}\").mkdir(parents=True, exist_ok=True)       \n",
    "\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/Cell_types_files_v2.csv'\n",
    "#load the genome info\n",
    "genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa.sizes\"\n",
    "genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa\"\n",
    "#genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/genomes/GRCh38.p13/GRCh38.p13.fa.sizes\"\n",
    "#genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/genomes/GRCh38.p13/GRCh38.p13.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>Accesibility_peakfiles</th>\n",
       "      <th>Merged_peakfiles</th>\n",
       "      <th>scATAC_BAMfiles</th>\n",
       "      <th>TPM_matrix</th>\n",
       "      <th>compare_with</th>\n",
       "      <th>count_table_files</th>\n",
       "      <th>count_table_files_blob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CB_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CjS</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CjS_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CSB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CSB_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSSCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CSSCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LNPCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>LNPCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LPCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>LPCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>StC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>StC_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CB_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CjS_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CSB_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CSSCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>LNPCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LPCs_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>StC_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MEC_blobun</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CSB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cell_type                             Accesibility_peakfiles  \\\n",
       "0             CB  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1            CjS  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "2            CSB  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "3          CSSCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "4          LNPCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "5           LPCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "6            StC  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "7            MEC  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "8      CB_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "9     CjS_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "10    CSB_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "11  CSSCs_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "12  LNPCs_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "13   LPCs_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "14    StC_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "15    MEC_blobun  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "                                     Merged_peakfiles  \\\n",
       "0   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "1   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "2   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "3   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "4   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "5   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "6   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "7   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "8   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "9   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "10  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "11  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "12  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "13  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "14  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "15  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "\n",
       "                                      scATAC_BAMfiles  \\\n",
       "0   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "2   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "3   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "4   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "5   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "6   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "7   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "8   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "9   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "10  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "11  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "12  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "13  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "14  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "15  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "                                           TPM_matrix  compare_with  \\\n",
       "0   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...     CB_blobun   \n",
       "1   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...    CjS_blobun   \n",
       "2   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...    CSB_blobun   \n",
       "3   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...  CSSCs_blobun   \n",
       "4   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...  LNPCs_blobun   \n",
       "5   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...   LPCs_blobun   \n",
       "6   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...    StC_blobun   \n",
       "7   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...    MEC_blobun   \n",
       "8   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "9   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "10  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "11  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "12  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "13  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "14  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           MEC   \n",
       "15  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...           CSB   \n",
       "\n",
       "                                    count_table_files  \\\n",
       "0   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "1   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "2   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "3   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "4   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "5   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "6   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "7   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "8   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "9   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "10  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "11  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "12  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "13  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "14  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "15  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...   \n",
       "\n",
       "                               count_table_files_blob  \n",
       "0   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "1   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "2   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "3   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "4   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "5   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "6   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "7   /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/bl...  \n",
       "8                                                 NaN  \n",
       "9                                                 NaN  \n",
       "10                                                NaN  \n",
       "11                                                NaN  \n",
       "12                                                NaN  \n",
       "13                                                NaN  \n",
       "14                                                NaN  \n",
       "15                                                NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "#sample_data = sample_data.iloc[[0,1],:]\n",
    "\n",
    "#sample_data=sample_data[3:10] use this for smaller populations comparison\n",
    "\n",
    "sample_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets first perform motif enrichment\n",
    "if not os.path.exists(f\"{output_dir}scan_results_acc.scanfile\"): \n",
    "    sp.check_call(f'nice -15 gimme scan '\n",
    "        f'-T -N 4 '\n",
    "        f'-g hg38 '\n",
    "        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/merge_peaks/2021-27-06/tmp/accesible_summits.bed '\n",
    "        f'> {output_dir}scan_results_acc.scanfile',shell = True)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets perform motif enrichment for ESC data\n",
    "if not os.path.exists(f\"{output_dir}scan_results_ESC.scanfile\"): \n",
    "    sp.check_call(f'nice -15 gimme scan '\n",
    "        f'-T -N 4 '\n",
    "        f'-g hg38 '\n",
    "        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/all_peaks.tsv '\n",
    "        f'> {output_dir}scan_results_ESC.scanfile',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "CjS\n",
      "CSB\n",
      "CSSCs\n",
      "LNPCs\n",
      "LPCs\n",
      "StC\n",
      "MEC\n",
      "CB_blobun\n",
      "running ananse binding for celltype CB_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/CB_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "CjS_blobun\n",
      "running ananse binding for celltype CjS_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/CjS_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "CSB_blobun\n",
      "running ananse binding for celltype CSB_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/CSB_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "CSSCs_blobun\n",
      "running ananse binding for celltype CSSCs_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/CSSCs_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "LNPCs_blobun\n",
      "running ananse binding for celltype LNPCs_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/LNPCs_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "LPCs_blobun\n",
      "running ananse binding for celltype LPCs_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/LPCs_blob.bam with the ATAC merged peaks of all joined peaks\n",
      "StC_blobun\n",
      "running ananse binding for celltype StC_blobun using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/split_bam_blob/splitblob/outs/subsets/StC_blob.bam with the ATAC merged peaks of all joined peaks\n"
     ]
    }
   ],
   "source": [
    "# lets do ananse binding for your cell populations\n",
    "i = 0\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}\").mkdir(parents=True, exist_ok=True)       \n",
    "    bam_file = sample_data.iloc[index,3]\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/binding.h5\"): \n",
    "        print(f'running ananse binding for celltype {cell_id} using the scATAC bamfile {bam_file} with the ATAC merged peaks of all joined peaks')\n",
    "        #with open(f'{output_dir}/{cell_id}/myfile.txt', 'w') as fp: \n",
    "       #     pass\n",
    "        sp.check_call(f'ananse binding '\n",
    "                f'-g {genome_path} '\n",
    "                f'-A {bam_file} '\n",
    "                f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/merge_peaks/accesible_summits.bed '\n",
    "                f'-o {output_dir}/{cell_id} '\n",
    "                f'--pfmscorefile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outsscan_results_acc.scanfile '\n",
    "                f'-n 4 '    \n",
    "                f'2> {output_dir}/{cell_id}/ananse_binding_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generating the binding.h5 file for ESC data ananse 0.3.0\n",
    "Path(f\"{output_dir}/ESC\").mkdir(parents=True, exist_ok=True)       \n",
    "if not os.path.exists(f\"{output_dir}/ESC/binding.h5\"): \n",
    "    sp.check_call(f'ananse binding '\n",
    "        f'-g {genome_path} '\n",
    "        f'-A /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/H3K27ac/results/final_bam/hg38-GSM466732.samtools-coordinate.bam '\n",
    "        f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/all_peaks.tsv '\n",
    "        f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC '\n",
    "        f'--pfmscorefile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3scan_results_ESC.scanfile '\n",
    "        f'-n 4 '   \n",
    "        f'2> {output_dir}/ESC/ananse_binding_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse network\n",
    "\n",
    "Generate GRN's based on the binding.tsv output and RNAseq expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make one TPM file per sample using the big count matrix containing all samples; already done\n",
    "# ensembl2hugo =  pd.read_table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/genomepy_genomes/GRCh38.p13/gene_id2name.tsv',sep = '\\t', comment ='_',names = ['ensembl','HUGO'])\n",
    "# for index, cell_type in sample_data.iterrows():\n",
    "#     for binding_mode in ['only_H3K27ac','Remap_H3K27ac']:\n",
    "#         cell_id = sample_data.iloc[index,0]\n",
    "#         TPM_matrix_file = sample_data.TPM_matrix[index]\n",
    "#         TPM_matrix = pd.read_table(TPM_matrix_file,sep = '\\t',index_col = 0)\n",
    "        \n",
    "#         #convert matrix to gene name; not nessesary if you do it beforehand\n",
    "#         #hugo_tpm_matrix = TPM_matrix.merge(ensembl2hugo, left_index = True, right_on = 'ensembl', how = 'inner')\n",
    "#         #hugo_tpm_matrix.index = hugo_tpm_matrix.HUGO\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop_duplicates()\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop_duplicates(subset='HUGO', keep=\"first\")\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop(['ensembl','HUGO'], axis = 1)\n",
    "\n",
    "#         TPM_files = sample_data.TPM_names[index].split(' ')\n",
    "#         Path(f\"{output_dir}/{binding_mode}/{cell_id}/TPM_files/\").mkdir(parents=True, exist_ok=True) \n",
    "#         for sample in TPM_files:\n",
    "#             sample_matrix = hugo_tpm_matrix[sample]\n",
    "#             sample_matrix = pd.DataFrame(sample_matrix)\n",
    "#             sample_matrix[['target_id']] = sample_matrix.index\n",
    "#             sample_matrix[['tpm']] = sample_matrix[[sample]]\n",
    "#             sample_matrix[['target_id','tpm']]\n",
    "#             #save sample TPM file\n",
    "#             sample_matrix[['target_id','tpm']].to_csv(f\"{output_dir}/{binding_mode}/{cell_id}/TPM_files/FPM_{sample}.tsv\", sep = '\\t', index = False, header = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control_t0B</th>\n",
       "      <th>control_t0A</th>\n",
       "      <th>tcdd_t2A</th>\n",
       "      <th>tcdd_t2B</th>\n",
       "      <th>control_t24A</th>\n",
       "      <th>control_t24B</th>\n",
       "      <th>tcdd_t24A</th>\n",
       "      <th>tcdd_t24B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUGO</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>17.149720</td>\n",
       "      <td>10.471415</td>\n",
       "      <td>10.595141</td>\n",
       "      <td>7.870116</td>\n",
       "      <td>6.327820</td>\n",
       "      <td>7.063620</td>\n",
       "      <td>7.979823</td>\n",
       "      <td>6.325595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>142.842093</td>\n",
       "      <td>107.685809</td>\n",
       "      <td>69.834197</td>\n",
       "      <td>68.042334</td>\n",
       "      <td>55.006878</td>\n",
       "      <td>53.782104</td>\n",
       "      <td>49.323561</td>\n",
       "      <td>41.976615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>4.579877</td>\n",
       "      <td>3.434859</td>\n",
       "      <td>2.408801</td>\n",
       "      <td>1.757298</td>\n",
       "      <td>2.696829</td>\n",
       "      <td>2.355813</td>\n",
       "      <td>2.560131</td>\n",
       "      <td>2.280325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>24.530547</td>\n",
       "      <td>17.480404</td>\n",
       "      <td>16.068732</td>\n",
       "      <td>11.890818</td>\n",
       "      <td>1.880808</td>\n",
       "      <td>1.570157</td>\n",
       "      <td>6.456277</td>\n",
       "      <td>6.035803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC064824.1</th>\n",
       "      <td>0.056404</td>\n",
       "      <td>0.107946</td>\n",
       "      <td>0.085177</td>\n",
       "      <td>0.021252</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016335</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.032467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL136225.2</th>\n",
       "      <td>0.056168</td>\n",
       "      <td>0.135234</td>\n",
       "      <td>0.090534</td>\n",
       "      <td>0.037190</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115257</td>\n",
       "      <td>0.082025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004636.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007687.1</th>\n",
       "      <td>0.355343</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.238526</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.237341</td>\n",
       "      <td>0.425962</td>\n",
       "      <td>0.032896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL353795.4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.571044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59143 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            control_t0B  control_t0A   tcdd_t2A   tcdd_t2B  control_t24A  \\\n",
       "HUGO                                                                       \n",
       "TSPAN6        17.149720    10.471415  10.595141   7.870116      6.327820   \n",
       "TNMD           0.000000     0.000000   0.000000   0.000000      0.000000   \n",
       "DPM1         142.842093   107.685809  69.834197  68.042334     55.006878   \n",
       "SCYL3          4.579877     3.434859   2.408801   1.757298      2.696829   \n",
       "C1orf112      24.530547    17.480404  16.068732  11.890818      1.880808   \n",
       "...                 ...          ...        ...        ...           ...   \n",
       "AC064824.1     0.056404     0.107946   0.085177   0.021252      0.000000   \n",
       "AL136225.2     0.056168     0.135234   0.090534   0.037190      0.000000   \n",
       "AC004636.1     0.000000     0.000000   0.000000   0.000000      0.000000   \n",
       "AC007687.1     0.355343     0.000000   0.000000   0.238526      0.000000   \n",
       "AL353795.4     0.000000     0.000000   0.000000   0.571044      0.000000   \n",
       "\n",
       "            control_t24B  tcdd_t24A  tcdd_t24B  \n",
       "HUGO                                            \n",
       "TSPAN6          7.063620   7.979823   6.325595  \n",
       "TNMD            0.000000   0.000000   0.000000  \n",
       "DPM1           53.782104  49.323561  41.976615  \n",
       "SCYL3           2.355813   2.560131   2.280325  \n",
       "C1orf112        1.570157   6.456277   6.035803  \n",
       "...                  ...        ...        ...  \n",
       "AC064824.1      0.016335   0.000000   0.032467  \n",
       "AL136225.2      0.000000   0.115257   0.082025  \n",
       "AC004636.1      0.000000   0.000000   0.000000  \n",
       "AC007687.1      0.237341   0.425962   0.032896  \n",
       "AL353795.4      0.000000   0.000000   0.000000  \n",
       "\n",
       "[59143 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#hugo_tpm_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CBtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CjS\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CjStpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSB\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSB/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSBtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSSCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSSCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSSCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LNPCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LNPCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LNPCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LPCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LPCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LPCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "StC\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/StC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/StCtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "MEC\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/MEC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/MECtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CB_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CBblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CjS_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CjSblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSB_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSB_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSBblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSSCs_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSSCs_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSSCsblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LNPCs_blobun#############################] | 100% Completed | 44.5s\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LNPCs_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LNPCsblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LPCs_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LPCs_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LPCsblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "StC_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/StC_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/StCblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "MEC_blobun\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/MEC_blobun/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/MECblobtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n"
     ]
    }
   ],
   "source": [
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    binding_file = f\"{output_dir}/{cell_id}/binding.h5\"\n",
    "    print(f'running ananse binding using the enhancer file {binding_file}')\n",
    "    file_name = sample_data.iloc[index,4]\n",
    "    print(f'together with the TPM file {file_name}')\n",
    "    print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"): \n",
    "        sp.check_call(f'nice -15 ananse network '\n",
    "            f'-e {file_name} '\n",
    "            f'-b {binding_file} '\n",
    "            f'-o {output_dir}/{cell_id}/full_network_includeprom.txt '\n",
    "            f'-g hg38 '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_network_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/tpm_ESC.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "[########################################] | 100% Completed | 52.2s\r"
     ]
    }
   ],
   "source": [
    "# generate the ESC network file from the binding.h5 file\n",
    "binding_file = f\"{output_dir}/ESC/binding.h5\"\n",
    "print(f'running ananse binding using the enhancer file {binding_file}')\n",
    "file_name = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/tpm_ESC.tsv\"\n",
    "print(f'together with the TPM file {file_name}')\n",
    "print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "if not os.path.exists(f\"{output_dir}/ESC/full_network_includeprom.txt\"): \n",
    "    sp.check_call(f'nice -15 ananse network '\n",
    "        f'-e {file_name} '\n",
    "        f'-b {binding_file} '\n",
    "        f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt '\n",
    "        f'-g hg38 '\n",
    "        f'-n 1 '\n",
    "        f'2> {output_dir}/ESC/ananse_network_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse Influence\n",
    "\n",
    "Lets finally run Ananse influence to predict key TFs that differ between cell types.\n",
    "it will compare each network with the networks listed in the 'comparison' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[43334, 43335, 43334, 43334]\n",
      "[42583, 42583, 42583, 42583]\n"
     ]
    }
   ],
   "source": [
    "# check threading and os delay\n",
    "import joblib\n",
    "from joblib import Parallel, delayed\n",
    "from joblib import parallel_backend\n",
    "\n",
    "print(Parallel(n_jobs=2)(delayed(os.getpid)() for i in range(4)))\n",
    "\n",
    "with parallel_backend('threading'):\n",
    "    print(Parallel(n_jobs=2)(delayed(os.getpid)() for i in range(4)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "100000\n"
     ]
    }
   ],
   "source": [
    "# adapting the recursion limit\n",
    "print(sys.getrecursionlimit())\n",
    "sys.setrecursionlimit(100000)\n",
    "print(sys.getrecursionlimit())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/peakpred_V4/only_ATAC/ESCs/full_network_include-promoter.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20210804/CBESCpseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/scratch/julian/tmp/z20m/ipykernel_5799/942299276.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{output_dir}/{cell_id}/ESC{cell_id}_influence_500000_V3/influence.txt\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m         sp.check_call(f'nice -15 ananse influence '\n\u001b[0m\u001b[1;32m     14\u001b[0m             \u001b[0;34mf'-s {net2} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m             \u001b[0;34mf'-t {network_file} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    366\u001b[0m     \u001b[0mcheck_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ls\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"-l\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    367\u001b[0m     \"\"\"\n\u001b[0;32m--> 368\u001b[0;31m     \u001b[0mretcode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    369\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mretcode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    370\u001b[0m         \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"args\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mcall\u001b[0;34m(timeout, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    349\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    350\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 351\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    352\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Including KeyboardInterrupt, wait handled that.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    353\u001b[0m             \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1187\u001b[0m             \u001b[0mendtime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_time\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1188\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1189\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1190\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1191\u001b[0m             \u001b[0;31m# https://bugs.python.org/issue25942\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36m_wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1915\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m                             \u001b[0;32mbreak\u001b[0m  \u001b[0;31m# Another thread waited.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m                         \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_try_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m                         \u001b[0;31m# Check the pid and loop as waitpid has been known to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m                         \u001b[0;31m# return 0 even without WNOHANG in odd situations.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36m_try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1873\u001b[0m             \u001b[0;34m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1874\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1875\u001b[0;31m                 \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwait_flags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1876\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mChildProcessError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1877\u001b[0m                 \u001b[0;31m# This happens if SIGCLD is set to be ignored or waiting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# cell population of interest against only ESC data\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/peakpred_V4/only_ATAC/ESCs/full_network_include-promoter.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/ESC{cell_id}_influence_500000_V3\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC{cell_id}_influence_500000_V3/influence.txt\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-s {net2} '\n",
    "            f'-t {network_file} '\n",
    "            f'-d {file_name} '\n",
    "            f'-i 500000 '\n",
    "            f'--plot '    \n",
    "            f'-o {output_dir}/{cell_id}/ESC{cell_id}_influence_500000_V3/influence.txt '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_V2.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB_blobun/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/blob20210804/CBblobpseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS_blobun/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/blob20210804/CjSblobpseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# Compare cell_types against blob of all other cells (normalized expression)\n",
    "for index, cell_type in sample_data.iterrows(): \n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    comparisons = sample_data.compare_with[index]\n",
    "    if not cell_id.find(\"blobun\") != -1:\n",
    "        #print(cell_id)\n",
    "        #print(comparisons)\n",
    "        network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "        print(f'running ananse influence using the network file {network_file}')\n",
    "        net2 = f\"{output_dir}/{comparisons}/full_network_includeprom.txt\"\n",
    "        print(f'compared to the network file {net2}')\n",
    "        Path(f\"{output_dir}/{comparisons}/{comparisons}_influence_500000_V3\").mkdir(parents=True, exist_ok=True) \n",
    "        file_name = sample_data.iloc[index,7]\n",
    "        print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "        if not os.path.exists(f\"{output_dir}/{comparisons}/{comparisons}_influence_500000_V3/influence_diffnetwork.txt\"):\n",
    "            sp.check_call(f'nice -15 ananse influence '\n",
    "                f'-s {net2} '\n",
    "                f'-t {network_file} '\n",
    "                f'-d {file_name} '\n",
    "                f'-i 500000 '\n",
    "                f'--plot '    \n",
    "                f'-o {output_dir}/{comparisons}/{comparisons}_influence_500000_V3/influence.txt '\n",
    "                f'-n 1 '\n",
    "                f'2> {output_dir}/{comparisons}/ananse_influence_log_V3_500.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extra code    \n",
    "    comparisons = sample_data.compare_with[index]\n",
    "        cell_id = sample_data.iloc[index,0]\n",
    "        #get the count data files for the source cell type\n",
    "        #count_files_source_cell_type = sample_data.TPM_names[index].split(' ')\n",
    "        if not comparisons != comparisons: #do stuff if there is a string in the comparisons column\n",
    "            for comparison in comparisons.split(' '):\n",
    "                #get the count data files for the target cell type\n",
    "                target_data = sample_data.loc[sample_data['cell_type'] == comparison]\n",
    "                target_data = target_data.reset_index(drop=True)\n",
    "                count_files_target_cell_type = target_data.TPM_names[0].split(' ')\n",
    "                #lets run deseq between the sample comparisons \n",
    "                comp_name = f'{cell_id}_to_{comparison}'\n",
    "                print(f'comparing {comp_name}')\n",
    "                if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}\"):\n",
    "                    Path(f\"{output_dir}/{binding_mode}/influence/{comp_name}\").mkdir(parents=True, exist_ok=True)\n",
    "                    \n",
    "                        source_network = f'{output_dir}/{cell_id}/full_network__includeprom.txt'\n",
    "                        target_network = f'{output_dir}/{comparison}/full_network_includeprom.txt'\n",
    "                        influence_output_name = f'{output_dir}/{binding_mode}/influence/{comp_name}/{weight_modus}/{n_edges}/'\n",
    "                        if not os.path.exists(f\"{influence_output_name}\"):\n",
    "                            Path(f\"{influence_output_name}\").mkdir(parents=True, exist_ok=True)\n",
    "                        if not os.path.exists(f\"{influence_output_name}influence.txt\"):\n",
    "                            print(f'running influence for {comp_name} with the datatype of {binding_mode} and {weight_modus}')\n",
    "                            sp.check_call(f'nice -15 ananse influence '\n",
    "                                f'-s {source_network} '\n",
    "                                f'-t {target_network} '\n",
    "                                f'-d {DEG_file} '\n",
    "                                f'--edges {n_edges} '\n",
    "                                f'-p '\n",
    "                                f'-o {influence_output_name}influence.txt ' \n",
    "                                f'-n 1 '\n",
    "                                f'2> {influence_output_name}log.txt',shell = True)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing KC_0h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_0h_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_24h_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_24h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_24h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_0and24_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0and24_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0and24_CTR with the datatype of Remap_H3K27ac and include-promoter\n"
     ]
    }
   ],
   "source": [
    "### delete in full??\n",
    "\n",
    "ensembl2hugo =  pd.read_table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/genomepy_genomes/GRCh38.p13/gene_id2name.tsv',sep = '\\t', comment ='_',names = ['ensembl','HUGO'])\n",
    "\n",
    "\n",
    "\n",
    "nice -15 ananse influence \n",
    "-s /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/peakpred_V4/only_ATAC/ESCs/full_network_include-promoter.txt \n",
    "-t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs/CjS/full_network_includeprom.txt \n",
    "-d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/pseudobulk_rnaseq/CjSESC_degenes.tsv \n",
    "-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs/CjS/ESCCjS_influence/influence.txt -n 10\n",
    "\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    for binding_mode in ['Remap_H3K27ac','only_H3K27ac']:\n",
    "        comparisons = sample_data.compare_with[index]\n",
    "        cell_id = sample_data.iloc[index,0]\n",
    "        #get the count data files for the source cell type\n",
    "        count_files_source_cell_type = sample_data.TPM_names[index].split(' ')\n",
    "        if not comparisons != comparisons: #do stuff if there is a string in the comparisons column\n",
    "            for comparison in comparisons.split(' '):\n",
    "                #get the count data files for the target cell type\n",
    "                target_data = sample_data.loc[sample_data['cell_type'] == comparison]\n",
    "                target_data = target_data.reset_index(drop=True)\n",
    "                count_files_target_cell_type = target_data.TPM_names[0].split(' ')\n",
    "                #lets run deseq between the sample comparisons \n",
    "                comp_name = f'{cell_id}_to_{comparison}'\n",
    "                print(f'comparing {comp_name}')\n",
    "                if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}\"):\n",
    "                    Path(f\"{output_dir}/{binding_mode}/influence/{comp_name}\").mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "                #lets load the countmatrix, and remove all columns that don't contain one of the samples\n",
    "                count_matrix = pd.read_table(sample_data.count_table_files[index])\n",
    "                #count_matrix[(count_files_source_cell_type + count_files_target_cell_type)]\n",
    "                #switch ensembl to HUGO gene names\n",
    "                hugo_count_matrix = count_matrix.merge(ensembl2hugo, left_on = 'gene', right_on = 'ensembl', how = 'inner')\n",
    "                hugo_count_matrix.index = hugo_count_matrix.HUGO\n",
    "                hugo_count_matrix = hugo_count_matrix.drop_duplicates()\n",
    "                hugo_count_matrix = hugo_count_matrix.drop_duplicates(subset='HUGO', keep=\"first\")\n",
    "                hugo_count_matrix = hugo_count_matrix.drop(['ensembl','HUGO','gene'], axis = 1)\n",
    "                #hugo_count_matrix.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv\", sep = '\\t', index = True, header = True)\n",
    "                hugo_count_matrix = hugo_count_matrix[(count_files_source_cell_type + count_files_target_cell_type)]\n",
    "                hugo_count_matrix.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv\", sep = '\\t', index = True, header = True)\n",
    "                #make a sample file for deseq2 \n",
    "                sample_names = (count_files_source_cell_type + count_files_target_cell_type)\n",
    "                annotation = [\"source\"] * len(count_files_source_cell_type) + [\"target\"] * len(count_files_target_cell_type)\n",
    "                data= {'samples':sample_names, 'condition':annotation}\n",
    "                sample_info_df = pd.DataFrame(data)\n",
    "                sample_info_df.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/sample_info.tsv\", sep = ',', index = False, header = True)\n",
    "                #run Deseq2 \n",
    "                #if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv\"):\n",
    "                \n",
    "                if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv\"):\n",
    "                    print('running Deseq2')\n",
    "                    sp.check_call(\n",
    "                        f'nice -5 Rscript --vanilla '\n",
    "                        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/RunDeseq2.R '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/sample_info.tsv '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv '\n",
    "                        f'2> {output_dir}/{binding_mode}/influence/{comp_name}/deseq2log.txt',\n",
    "                        shell = True)\n",
    "\n",
    "                    DEG_table = pd.read_table(f'{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv', \n",
    "                            sep = ',', comment = '#')\n",
    "                    DEG_table['resid'] = DEG_table['region']\n",
    "                    DEG_table = DEG_table.dropna()\n",
    "                    DEG_table[['resid','log2FoldChange','padj']].to_csv(f'{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv', sep = '\\t', index = False, header = True)\n",
    "                    DEG_file = f'{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv'\n",
    "\n",
    "                for weight_modus in ['include-promoter']:\n",
    "                    for n_edges in [100000, 500000]:\n",
    "                        source_network = f'{output_dir}/{binding_mode}/{cell_id}/full_network_{weight_modus}.txt'\n",
    "                        target_network = f'{output_dir}/{binding_mode}/{comparison}/full_network_{weight_modus}.txt'\n",
    "                        influence_output_name = f'{output_dir}/{binding_mode}/influence/{comp_name}/{weight_modus}/{n_edges}/'\n",
    "                        if not os.path.exists(f\"{influence_output_name}\"):\n",
    "                            Path(f\"{influence_output_name}\").mkdir(parents=True, exist_ok=True)\n",
    "                        if not os.path.exists(f\"{influence_output_name}influence.txt\"):\n",
    "                            print(f'running influence for {comp_name} with the datatype of {binding_mode} and {weight_modus}')\n",
    "                            sp.check_call(f'nice -15 ananse influence '\n",
    "                                f'-s {source_network} '\n",
    "                                f'-t {target_network} '\n",
    "                                f'-d {DEG_file} '\n",
    "                                f'-p '\n",
    "                                f'-o {influence_output_name}influence.txt ' \n",
    "                                f'--edges {n_edges} '\n",
    "                                f'-n 12 '\n",
    "                                f'2> {influence_output_name}log.txt',shell = True)  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
