{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask\n",
    "import dask.dataframe as dd\n",
    "from scipy import stats\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import shutil\n",
    "import glob\n",
    "import gimmemotifs\n",
    "from pathlib import Path\n",
    "import qnorm\n",
    "from sklearn import datasets, linear_model\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import mean_squared_error, r2_score, average_precision_score\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hue_regplot(data, x, y, hue, palette=None, **kwargs):\n",
    "    from matplotlib.cm import get_cmap\n",
    "    \n",
    "    regplots = []\n",
    "    \n",
    "    levels = data[hue].unique()\n",
    "    \n",
    "    if palette is None:\n",
    "        default_colors = get_cmap('tab10')\n",
    "        palette = {k: default_colors(i) for i, k in enumerate(levels)}\n",
    "    \n",
    "    for key in levels:\n",
    "        regplots.append(\n",
    "            sns.regplot(\n",
    "                x=x,\n",
    "                y=y,\n",
    "                fit_reg=False, \n",
    "                data=data[data[hue] == key],\n",
    "                color=palette[key],\n",
    "                **kwargs\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    return regplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3'\n",
    "Path(f\"{output_dir}\").mkdir(parents=True, exist_ok=True)       \n",
    "\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/Cell_types_files_v2.csv'\n",
    "#load the genome info\n",
    "genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa.sizes\"\n",
    "genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa\"\n",
    "#genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/genomes/GRCh38.p13/GRCh38.p13.fa.sizes\"\n",
    "#genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/genomes/GRCh38.p13/GRCh38.p13.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>Accesibility_peakfiles</th>\n",
       "      <th>Merged_peakfiles</th>\n",
       "      <th>scATAC_BAMfiles</th>\n",
       "      <th>TPM_matrix</th>\n",
       "      <th>compare_with</th>\n",
       "      <th>count_table_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CjS</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CSB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSSCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LNPCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LPCs</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>StC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MEC</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...</td>\n",
       "      <td>CSB</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_type                             Accesibility_peakfiles  \\\n",
       "0        CB  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1       CjS  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "2       CSB  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "3     CSSCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "4     LNPCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "5      LPCs  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "6       StC  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "7       MEC  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "                                    Merged_peakfiles  \\\n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "2  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "3  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "4  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "5  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "6  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "7  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...   \n",
       "\n",
       "                                     scATAC_BAMfiles  \\\n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "2  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "3  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "4  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "5  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "6  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "7  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "                                          TPM_matrix compare_with  \\\n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "2  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "3  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "4  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "5  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "6  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          MEC   \n",
       "7  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupy...          CSB   \n",
       "\n",
       "                                   count_table_files  \n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "2  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "3  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "4  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "5  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "6  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  \n",
       "7  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "#sample_data = sample_data.iloc[[0,1],:]\n",
    "\n",
    "#sample_data=sample_data[3:10] use this for smaller populations comparison\n",
    "\n",
    "sample_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets first perform motif enrichment\n",
    "if not os.path.exists(f\"{output_dir}scan_results_acc.scanfile\"): \n",
    "    sp.check_call(f'nice -15 gimme scan '\n",
    "        f'-T -N 4 '\n",
    "        f'-g hg38 '\n",
    "        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/merge_peaks/2021-27-06/tmp/accesible_summits.bed '\n",
    "        f'> {output_dir}scan_results_acc.scanfile',shell = True)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets perform motif enrichment for ESC data\n",
    "if not os.path.exists(f\"{output_dir}scan_results_ESC.scanfile\"): \n",
    "    sp.check_call(f'nice -15 gimme scan '\n",
    "        f'-T -N 4 '\n",
    "        f'-g hg38 '\n",
    "        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/all_peaks.tsv '\n",
    "        f'> {output_dir}scan_results_ESC.scanfile',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "CjS\n",
      "CSB\n",
      "running ananse binding for celltype CSB using the scATAC datatypes with the ATAC merged peaks of all joined peaks\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 14:56:53 | INFO | using default motif file\n",
      "2021-07-19 14:56:58 | INFO | using motifs for 828 factors\n",
      "2021-07-19 14:57:00 | INFO | loading pre-scanned motif scores.\n",
      "2021-07-19 14:58:08 | INFO | loading ATAC data\n",
      "2021-07-19 15:07:41 | DEBUG | quantile normalization for ATAC\n",
      "2021-07-19 15:07:43 | INFO | Loading models\n",
      "2021-07-19 15:07:44 | INFO | 238 models found\n",
      "2021-07-19 15:07:45 | INFO | Predicting TF activity\n",
      "2021-07-19 15:07:49,816 - INFO - motif scanning (scores)\n",
      "2021-07-19 15:07:49,817 - INFO - reading table\n",
      "2021-07-19 15:09:17,317 - INFO - creating score table (z-score, GC%)\n",
      "2021-07-19 15:12:48,874 - INFO - done\n",
      "2021-07-19 15:12:51,013 - INFO - creating dataframe\n",
      "2021-07-19 15:12:57,930 - INFO - Fitting BayesianRidge\n",
      "100%|██████████| 1/1 [00:15<00:00, 15.10s/it]\n",
      "2021-07-19 15:13:16,460 - INFO - Done\n",
      "2021-07-19 15:13:17 | INFO | Using SOX9 motif with SOX13 model weights\n",
      "2021-07-19 15:13:17 | INFO | Using SOX13 model\n",
      "2021-07-19 15:13:17 | INFO | Using SRY motif with SOX2 model weights\n",
      "2021-07-19 15:13:18 | INFO | Using SOX15 motif with SOX2 model weights\n",
      "2021-07-19 15:13:18 | INFO | Using SOX30 motif with SOX13 model weights\n",
      "2021-07-19 15:13:19 | INFO | Using SOX7 motif with SOX2 model weights\n",
      "2021-07-19 15:13:19 | INFO | Using SOX12 motif with SOX13 model weights\n",
      "2021-07-19 15:13:19 | INFO | Using SOX18 motif with SOX2 model weights\n",
      "2021-07-19 15:13:20 | INFO | Using TGIF1 motif with TGIF2 model weights\n",
      "2021-07-19 15:13:20 | INFO | Using SRF model\n",
      "2021-07-19 15:13:21 | INFO | Using EGR1 model\n",
      "2021-07-19 15:13:21 | INFO | Using RORA motif with RARA model weights\n",
      "2021-07-19 15:13:21 | INFO | Using NR1D1 motif with NR2C1 model weights\n",
      "2021-07-19 15:13:22 | INFO | Using RORC motif with NR2C1 model weights\n",
      "2021-07-19 15:13:22 | INFO | Using PAX4 motif with PDX1 model weights\n",
      "2021-07-19 15:13:22 | INFO | Using VSX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:13:23 | INFO | Using LBX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:13:23 | INFO | Using NR2C2 model\n",
      "2021-07-19 15:13:24 | INFO | Using HIF1A motif with AHR model weights\n",
      "2021-07-19 15:13:24 | INFO | Using AHR model\n",
      "2021-07-19 15:13:24 | INFO | Using ARNT model\n",
      "2021-07-19 15:13:25 | INFO | Using EPAS1 motif with AHR model weights\n",
      "2021-07-19 15:13:25 | INFO | Using SIM1 motif with AHR model weights\n",
      "2021-07-19 15:13:26 | INFO | Using NPAS3 motif with AHR model weights\n",
      "2021-07-19 15:13:26 | INFO | Using NPAS1 motif with AHR model weights\n",
      "2021-07-19 15:13:26 | INFO | Using MYBL2 model\n",
      "2021-07-19 15:13:27 | INFO | Using MYBL1 motif with MYBL2 model weights\n",
      "2021-07-19 15:13:27 | INFO | Using EGR2 motif with EGR1 model weights\n",
      "2021-07-19 15:13:27 | INFO | Using EGR4 motif with EGR1 model weights\n",
      "2021-07-19 15:13:28 | INFO | Using GATA1 model\n",
      "2021-07-19 15:13:28 | INFO | Using TAL1 model\n",
      "2021-07-19 15:13:29 | INFO | Using GATA2 model\n",
      "2021-07-19 15:13:29 | INFO | Using SNAI2 motif with ZEB1 model weights\n",
      "2021-07-19 15:13:29 | INFO | Using ZEB1 model\n",
      "2021-07-19 15:13:30 | INFO | Using SNAI1 motif with TGIF2 model weights\n",
      "2021-07-19 15:13:30 | INFO | Using SNAI3 motif with TGIF2 model weights\n",
      "2021-07-19 15:13:31 | INFO | Using NHLH2 motif with TAL1 model weights\n",
      "2021-07-19 15:13:31 | INFO | Using NHLH1 motif with TAL1 model weights\n",
      "2021-07-19 15:13:32 | INFO | Using CEBPB model\n",
      "2021-07-19 15:13:32 | INFO | Using DLX5 motif with SP5 model weights\n",
      "2021-07-19 15:13:32 | INFO | Using PRRX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:13:33 | INFO | Using ISX motif with MIXL1 model weights\n",
      "2021-07-19 15:13:33 | INFO | Using DLX2 motif with PDX1 model weights\n",
      "2021-07-19 15:13:33 | INFO | Using CDX4 motif with MIXL1 model weights\n",
      "2021-07-19 15:13:34 | INFO | Using OTP motif with MIXL1 model weights\n",
      "2021-07-19 15:13:34 | INFO | Using MSX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:13:35 | INFO | Using DLX6 motif with PDX1 model weights\n",
      "2021-07-19 15:13:35 | INFO | Using DLX3 motif with PDX1 model weights\n",
      "2021-07-19 15:13:35 | INFO | Using DLX4 motif with PDX1 model weights\n",
      "2021-07-19 15:13:36 | INFO | Using SPI1 model\n",
      "2021-07-19 15:13:36 | INFO | Using FEV motif with ELK1 model weights\n",
      "2021-07-19 15:13:37 | INFO | Using ETS2 motif with ELK1 model weights\n",
      "2021-07-19 15:13:37 | INFO | Using ETV2 motif with ELK1 model weights\n",
      "2021-07-19 15:13:37 | INFO | Using ERF motif with ELK1 model weights\n",
      "2021-07-19 15:13:38 | INFO | Using FOSL1 model\n",
      "2021-07-19 15:13:38 | INFO | Using JUND model\n",
      "2021-07-19 15:13:38 | INFO | Using JUNB model\n",
      "2021-07-19 15:13:39 | INFO | Using JDP2 motif with FOSL1 model weights\n",
      "2021-07-19 15:13:39 | INFO | Using FOS model\n",
      "2021-07-19 15:13:40 | INFO | Using FOSB motif with FOS model weights\n",
      "2021-07-19 15:13:40 | INFO | Using IRF1 model\n",
      "2021-07-19 15:13:40 | INFO | Using IRF3 model\n",
      "2021-07-19 15:13:41 | INFO | Using IRF2 model\n",
      "2021-07-19 15:13:41 | INFO | Using TCF3 model\n",
      "2021-07-19 15:13:42 | INFO | Using ATOH7 motif with NEUROD1 model weights\n",
      "2021-07-19 15:13:42 | INFO | Using RARG motif with RARA model weights\n",
      "2021-07-19 15:13:42 | INFO | Using RXRG motif with RXRB model weights\n",
      "2021-07-19 15:13:43 | INFO | Using RXRA model\n",
      "2021-07-19 15:13:43 | INFO | Using RARA model\n",
      "2021-07-19 15:13:43 | INFO | Using BACH1 model\n",
      "2021-07-19 15:13:44 | INFO | Using BACH2 motif with BACH1 model weights\n",
      "2021-07-19 15:13:44 | INFO | Using MAFK model\n",
      "2021-07-19 15:13:45 | INFO | Using NFE2 model\n",
      "2021-07-19 15:13:45 | INFO | Using SREBF1 model\n",
      "2021-07-19 15:13:45 | INFO | Using HSF1 model\n",
      "2021-07-19 15:13:46 | INFO | Using HSF4 motif with HSF1 model weights\n",
      "2021-07-19 15:13:46 | INFO | Using HOXB7 motif with PDX1 model weights\n",
      "2021-07-19 15:13:47 | INFO | Using HOXB6 motif with PDX1 model weights\n",
      "2021-07-19 15:13:47 | INFO | Using HOXA5 motif with PDX1 model weights\n",
      "2021-07-19 15:13:47 | INFO | Using HOXB8 motif with PDX1 model weights\n",
      "2021-07-19 15:13:48 | INFO | Using HOXB4 motif with PDX1 model weights\n",
      "2021-07-19 15:13:48 | INFO | Using HOXB9 motif with MEIS2 model weights\n",
      "2021-07-19 15:13:48 | INFO | Using HOXD3 motif with PDX1 model weights\n",
      "2021-07-19 15:13:49 | INFO | Using HOXC9 motif with MEIS2 model weights\n",
      "2021-07-19 15:13:49 | INFO | Using HOXD4 motif with PDX1 model weights\n",
      "2021-07-19 15:13:50 | INFO | Using HOXC5 motif with PDX1 model weights\n",
      "2021-07-19 15:13:50 | INFO | Using HOXB2 motif with PDX1 model weights\n",
      "2021-07-19 15:13:50 | INFO | Using HOXA6 motif with PDX1 model weights\n",
      "2021-07-19 15:13:51 | INFO | No match for ZBTB6 based on motifs\n",
      "2021-07-19 15:13:51 | INFO | No related TF found for ZBTB6, using general model\n",
      "2021-07-19 15:13:51 | INFO | Using POU3F2 motif with POU5F1 model weights\n",
      "2021-07-19 15:13:52 | INFO | Using NANOG model\n",
      "2021-07-19 15:13:52 | INFO | Using POU5F1 model\n",
      "2021-07-19 15:13:52 | INFO | Using ZIC1 motif with ZBTB7B model weights\n",
      "2021-07-19 15:13:53 | INFO | Using ZIC4 motif with ZBTB7B model weights\n",
      "2021-07-19 15:13:53 | INFO | Using ZIC5 motif with ZBTB7B model weights\n",
      "2021-07-19 15:13:53 | INFO | No match for IRX3 based on motifs\n",
      "2021-07-19 15:13:53 | INFO | No related TF found for IRX3, using general model\n",
      "2021-07-19 15:13:54 | INFO | No match for IRX1 based on motifs\n",
      "2021-07-19 15:13:54 | INFO | No related TF found for IRX1, using general model\n",
      "2021-07-19 15:13:54 | INFO | No match for IRX6 based on motifs\n",
      "2021-07-19 15:13:54 | INFO | No related TF found for IRX6, using general model\n",
      "2021-07-19 15:13:55 | INFO | No match for IRX4 based on motifs\n",
      "2021-07-19 15:13:55 | INFO | No related TF found for IRX4, using general model\n",
      "2021-07-19 15:13:55 | INFO | Using NR2C1 model\n",
      "2021-07-19 15:13:55 | INFO | No match for DMRT1 based on motifs\n",
      "2021-07-19 15:13:55 | INFO | No related TF found for DMRT1, using general model\n",
      "2021-07-19 15:13:56 | INFO | Using TFE3 model\n",
      "2021-07-19 15:13:56 | INFO | Using USF2 model\n",
      "2021-07-19 15:13:57 | INFO | Using MITF model\n",
      "2021-07-19 15:13:57 | INFO | Using USF1 model\n",
      "2021-07-19 15:13:57 | INFO | Using CEBPG model\n",
      "2021-07-19 15:13:58 | INFO | Using ATF4 model\n",
      "2021-07-19 15:13:58 | INFO | Using DDIT3 motif with ATF4 model weights\n",
      "2021-07-19 15:13:59 | INFO | Using ATF3 model\n",
      "2021-07-19 15:13:59 | INFO | Using ATF5 motif with ATF4 model weights\n",
      "2021-07-19 15:13:59 | INFO | Using MYC model\n",
      "2021-07-19 15:14:00 | INFO | Using TBX20 motif with EOMES model weights\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:14:00 | INFO | Using TBX1 motif with TBX21 model weights\n",
      "2021-07-19 15:14:01 | INFO | Using EOMES model\n",
      "2021-07-19 15:14:01 | INFO | Using TBX21 model\n",
      "2021-07-19 15:14:01 | INFO | Using TBR1 motif with EOMES model weights\n",
      "2021-07-19 15:14:02 | INFO | Using TBX10 motif with TBX21 model weights\n",
      "2021-07-19 15:14:02 | INFO | Using MYF6 motif with TCF3 model weights\n",
      "2021-07-19 15:14:03 | INFO | Using FIGLA motif with TCF3 model weights\n",
      "2021-07-19 15:14:03 | INFO | Using MYF5 motif with TCF3 model weights\n",
      "2021-07-19 15:14:04 | INFO | Using PITX3 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:04 | INFO | Using OTX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:04 | INFO | Using CRX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:05 | INFO | Using DMBX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:05 | INFO | Using PITX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:05 | INFO | Using OTX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:06 | INFO | Using GSC2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:06 | INFO | Using GSC motif with MIXL1 model weights\n",
      "2021-07-19 15:14:07 | INFO | Using PITX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:07 | INFO | Using DPRX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:07 | INFO | Using GATA3 model\n",
      "2021-07-19 15:14:08 | INFO | Using RFX5 model\n",
      "2021-07-19 15:14:08 | INFO | Using RFX8 motif with RFX5 model weights\n",
      "2021-07-19 15:14:09 | INFO | Using RFX6 motif with RFX5 model weights\n",
      "2021-07-19 15:14:09 | INFO | Using RFX7 motif with RFX5 model weights\n",
      "2021-07-19 15:14:09 | INFO | Using RFX3 model\n",
      "2021-07-19 15:14:10 | INFO | Using RFX4 motif with RFX3 model weights\n",
      "2021-07-19 15:14:10 | INFO | Using NEUROD2 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:11 | INFO | Using NEUROD1 model\n",
      "2021-07-19 15:14:11 | INFO | Using BHLHA15 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:11 | INFO | Using ATOH1 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:12 | INFO | Using PTF1A motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:12 | INFO | Using TCF21 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:13 | INFO | Using NEUROG2 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:13 | INFO | Using OLIG2 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:13 | INFO | No match for DMRT2 based on motifs\n",
      "2021-07-19 15:14:13 | INFO | No related TF found for DMRT2, using general model\n",
      "2021-07-19 15:14:14 | INFO | No match for DMRTA1 based on motifs\n",
      "2021-07-19 15:14:14 | INFO | No related TF found for DMRTA1, using general model\n",
      "2021-07-19 15:14:14 | INFO | No match for DMRT3 based on motifs\n",
      "2021-07-19 15:14:14 | INFO | No related TF found for DMRT3, using general model\n",
      "2021-07-19 15:14:15 | INFO | Using NEUROD6 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:15 | INFO | Using NEUROD4 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:15 | INFO | Using TCF23 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:16 | INFO | Using NEUROG3 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:16 | INFO | Using TCF15 motif with TAL1 model weights\n",
      "2021-07-19 15:14:17 | INFO | Using NEUROG1 motif with NEUROD1 model weights\n",
      "2021-07-19 15:14:17 | INFO | Using NFKB1 motif with RELA model weights\n",
      "2021-07-19 15:14:17 | INFO | Using ELF3 model\n",
      "2021-07-19 15:14:18 | INFO | Using EHF motif with ELF3 model weights\n",
      "2021-07-19 15:14:18 | INFO | Using ELF5 motif with ETV5 model weights\n",
      "2021-07-19 15:14:18 | INFO | Using ELF1 model\n",
      "2021-07-19 15:14:19 | INFO | Using PAX8 model\n",
      "2021-07-19 15:14:19 | INFO | Using PAX5 model\n",
      "2021-07-19 15:14:20 | INFO | Using PAX9 motif with PAX8 model weights\n",
      "2021-07-19 15:14:20 | INFO | Using PAX1 motif with PAX8 model weights\n",
      "2021-07-19 15:14:20 | INFO | Using SOX4 motif with SOX6 model weights\n",
      "2021-07-19 15:14:21 | INFO | Using SOX6 model\n",
      "2021-07-19 15:14:21 | INFO | Using SOX17 motif with SOX6 model weights\n",
      "2021-07-19 15:14:21 | INFO | Using SOX2 model\n",
      "2021-07-19 15:14:22 | INFO | Using SOX3 motif with SOX2 model weights\n",
      "2021-07-19 15:14:22 | INFO | Using FOXI1 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:23 | INFO | Using FOXQ1 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:23 | INFO | Using FOXI3 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:23 | INFO | Using FOXD2 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:24 | INFO | Using FOXD4 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:24 | INFO | Using FOXE3 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:25 | INFO | Using FOXD4L5 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:25 | INFO | Using FOXE1 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:25 | INFO | Using FOXD3 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:26 | INFO | Using FOXB2 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:26 | INFO | Using FOXK2 model\n",
      "2021-07-19 15:14:27 | INFO | Using FOXI2 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:27 | INFO | Using FOXD4L3 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:27 | INFO | Using FOXF2 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:28 | INFO | Using FOXS1 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:28 | INFO | Using FOXD4L1 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:29 | INFO | Using FOXA3 model\n",
      "2021-07-19 15:14:29 | INFO | Using FOXA1 model\n",
      "2021-07-19 15:14:29 | INFO | Using FOXD4L6 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:30 | INFO | Using FOXL2 motif with FOXA3 model weights\n",
      "2021-07-19 15:14:30 | INFO | Using CLOCK model\n",
      "2021-07-19 15:14:30 | INFO | Using MAX model\n",
      "2021-07-19 15:14:31 | INFO | Using MXI1 model\n",
      "2021-07-19 15:14:31 | INFO | Using MYCN motif with CLOCK model weights\n",
      "2021-07-19 15:14:32 | INFO | Using NPAS2 motif with CLOCK model weights\n",
      "2021-07-19 15:14:32 | INFO | Using RAX2 motif with PDX1 model weights\n",
      "2021-07-19 15:14:32 | INFO | Using EN2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:33 | INFO | Using EN1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:33 | INFO | Using NOBOX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:33 | INFO | Using VSX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:34 | INFO | Using GBX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:34 | INFO | Using HESX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:35 | INFO | Using SHOX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:35 | INFO | Using LHX9 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:36 | INFO | Using RAX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:36 | INFO | Using LMX1B motif with PDX1 model weights\n",
      "2021-07-19 15:14:36 | INFO | Using UNCX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:37 | INFO | Using ALX4 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:37 | INFO | Using ALX3 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:37 | INFO | Using HOXA1 motif with PDX1 model weights\n",
      "2021-07-19 15:14:38 | INFO | Using LMX1A motif with MIXL1 model weights\n",
      "2021-07-19 15:14:38 | INFO | Using GBX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:39 | INFO | Using ALX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:39 | INFO | Using LHX4 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:39 | INFO | Using DLX1 motif with PDX1 model weights\n",
      "2021-07-19 15:14:40 | INFO | Using MSX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:40 | INFO | Using PRRX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:40 | INFO | Using LHX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:41 | INFO | Using SHOX motif with PDX1 model weights\n",
      "2021-07-19 15:14:41 | INFO | Using LHX5 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:42 | INFO | Using ARX motif with MIXL1 model weights\n",
      "2021-07-19 15:14:42 | INFO | Using ESX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:42 | INFO | Using LHX3 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:43 | INFO | Using ZFHX2 motif with PDX1 model weights\n",
      "2021-07-19 15:14:43 | INFO | Using HOXD1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:44 | INFO | Using HOXB1 motif with MIXL1 model weights\n",
      "2021-07-19 15:14:44 | INFO | Using PBX3 model\n",
      "2021-07-19 15:14:44 | INFO | Using PBX2 model\n",
      "2021-07-19 15:14:45 | INFO | Using PBX4 motif with PBX2 model weights\n",
      "2021-07-19 15:14:45 | INFO | Using SMAD4 model\n",
      "2021-07-19 15:14:46 | INFO | Using SMAD2 model\n",
      "2021-07-19 15:14:46 | INFO | Using SMAD3 model\n",
      "2021-07-19 15:14:46 | INFO | Using ELK3 motif with ETV1 model weights\n",
      "2021-07-19 15:14:47 | INFO | Using ETS1 model\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:14:47 | INFO | Using ETV4 model\n",
      "2021-07-19 15:14:47 | INFO | Using ELK4 model\n",
      "2021-07-19 15:14:48 | INFO | Using ELK1 model\n",
      "2021-07-19 15:14:48 | INFO | Using ETV1 model\n",
      "2021-07-19 15:14:49 | INFO | Using GABPA model\n",
      "2021-07-19 15:14:49 | INFO | Using ETV3 motif with ELK4 model weights\n",
      "2021-07-19 15:14:49 | INFO | Using ERG model\n",
      "2021-07-19 15:14:50 | INFO | Using FLI1 motif with ERG model weights\n",
      "2021-07-19 15:14:50 | INFO | Using NR4A2 motif with NR4A1 model weights\n",
      "2021-07-19 15:14:51 | INFO | Using NR1H4 motif with RXRA model weights\n",
      "2021-07-19 15:14:51 | INFO | Using NR1H3 motif with RXRB model weights\n",
      "2021-07-19 15:14:51 | INFO | Using SCRT1 motif with ZEB1 model weights\n",
      "2021-07-19 15:14:52 | INFO | Using ASCL2 motif with TCF12 model weights\n",
      "2021-07-19 15:14:52 | INFO | Using RORB motif with ESR1 model weights\n",
      "2021-07-19 15:14:52 | INFO | Using NR1D2 motif with NR2C1 model weights\n",
      "2021-07-19 15:14:53 | INFO | Using ZBTB7A model\n",
      "2021-07-19 15:14:54 | INFO | Using E2F2 motif with E2F1 model weights\n",
      "2021-07-19 15:14:54 | INFO | Using E2F4 model\n",
      "2021-07-19 15:14:54 | INFO | Using E2F1 model\n",
      "2021-07-19 15:14:55 | INFO | Using E2F3 motif with E2F5 model weights\n",
      "2021-07-19 15:14:55 | INFO | Using E2F5 model\n",
      "2021-07-19 15:14:55 | INFO | Using NR2E1 motif with NFIB model weights\n",
      "2021-07-19 15:14:56 | INFO | Using RUNX3 model\n",
      "2021-07-19 15:14:56 | INFO | Using RUNX1 model\n",
      "2021-07-19 15:14:57 | INFO | Using RUNX2 motif with RUNX3 model weights\n",
      "2021-07-19 15:14:57 | INFO | Using SP3 motif with SP5 model weights\n",
      "2021-07-19 15:14:57 | INFO | Using SP1 model\n",
      "2021-07-19 15:14:58 | INFO | Using SP2 model\n",
      "2021-07-19 15:14:58 | INFO | Using KLF4 model\n",
      "2021-07-19 15:14:58 | INFO | Using KLF5 model\n",
      "2021-07-19 15:14:59 | INFO | Using KLF12 motif with KLF5 model weights\n",
      "2021-07-19 15:14:59 | INFO | Using SP6 motif with SP5 model weights\n",
      "2021-07-19 15:15:00 | INFO | Using SP8 motif with SP5 model weights\n",
      "2021-07-19 15:15:00 | INFO | Using SP5 model\n",
      "2021-07-19 15:15:00 | INFO | Using KLF7 motif with KLF5 model weights\n",
      "2021-07-19 15:15:01 | INFO | Using SP9 motif with SP5 model weights\n",
      "2021-07-19 15:15:01 | INFO | Using KLF6 model\n",
      "2021-07-19 15:15:01 | INFO | Using KLF3 motif with KLF5 model weights\n",
      "2021-07-19 15:15:02 | INFO | Using THAP1 model\n",
      "2021-07-19 15:15:02 | INFO | Using SP7 motif with SP5 model weights\n",
      "2021-07-19 15:15:03 | INFO | Using KLF2 motif with KLF5 model weights\n",
      "2021-07-19 15:15:03 | INFO | Using GMEB2 model\n",
      "2021-07-19 15:15:03 | INFO | Using GMEB1 model\n",
      "2021-07-19 15:15:04 | INFO | Using SIX3 motif with GATA2 model weights\n",
      "2021-07-19 15:15:04 | INFO | Using SIX1 motif with GATA2 model weights\n",
      "2021-07-19 15:15:05 | INFO | Using SIX6 motif with GATA2 model weights\n",
      "2021-07-19 15:15:05 | INFO | Using SIX2 motif with GATA2 model weights\n",
      "2021-07-19 15:15:05 | INFO | No match for GCM1 based on motifs\n",
      "2021-07-19 15:15:05 | INFO | No related TF found for GCM1, using general model\n",
      "2021-07-19 15:15:06 | INFO | No match for GCM2 based on motifs\n",
      "2021-07-19 15:15:06 | INFO | No related TF found for GCM2, using general model\n",
      "2021-07-19 15:15:06 | INFO | Using POU4F2 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:07 | INFO | Using POU4F1 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:07 | INFO | Using POU3F3 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:07 | INFO | Using POU4F3 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:08 | INFO | Using POU1F1 motif with POU5F1 model weights\n",
      "2021-07-19 15:15:08 | INFO | Using POU3F1 motif with POU5F1 model weights\n",
      "2021-07-19 15:15:09 | INFO | Using POU3F4 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:09 | INFO | Using NFE2L2 model\n",
      "2021-07-19 15:15:09 | INFO | Using MAF motif with MAFK model weights\n",
      "2021-07-19 15:15:10 | INFO | Using HOXC13 motif with PDX1 model weights\n",
      "2021-07-19 15:15:10 | INFO | Using HOXD13 motif with PDX1 model weights\n",
      "2021-07-19 15:15:11 | INFO | Using HOXB13 motif with PDX1 model weights\n",
      "2021-07-19 15:15:11 | INFO | Using HOXA13 motif with PDX1 model weights\n",
      "2021-07-19 15:15:11 | INFO | Using HOXD10 motif with PDX1 model weights\n",
      "2021-07-19 15:15:12 | INFO | Using SOX5 model\n",
      "2021-07-19 15:15:12 | INFO | Using SOX11 motif with SOX13 model weights\n",
      "2021-07-19 15:15:12 | INFO | Using SOX8 motif with SOX5 model weights\n",
      "2021-07-19 15:15:13 | INFO | Using NFIL3 model\n",
      "2021-07-19 15:15:13 | INFO | Using HLF model\n",
      "2021-07-19 15:15:14 | INFO | Using HNF4G model\n",
      "2021-07-19 15:15:14 | INFO | Using HNF4A model\n",
      "2021-07-19 15:15:14 | INFO | No match for PLAGL1 based on motifs\n",
      "2021-07-19 15:15:14 | INFO | No related TF found for PLAGL1, using general model\n",
      "2021-07-19 15:15:15 | INFO | No match for GLIS3 based on motifs\n",
      "2021-07-19 15:15:15 | INFO | No related TF found for GLIS3, using general model\n",
      "2021-07-19 15:15:15 | INFO | No match for RHOXF1 based on motifs\n",
      "2021-07-19 15:15:15 | INFO | No related TF found for RHOXF1, using general model\n",
      "2021-07-19 15:15:16 | INFO | No match for SP140 based on motifs\n",
      "2021-07-19 15:15:16 | INFO | No related TF found for SP140, using general model\n",
      "2021-07-19 15:15:16 | INFO | Using POU2F1 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:17 | INFO | Using HOXB5 motif with PDX1 model weights\n",
      "2021-07-19 15:15:17 | INFO | Using HOXC8 motif with PDX1 model weights\n",
      "2021-07-19 15:15:17 | INFO | Using BSX motif with MIXL1 model weights\n",
      "2021-07-19 15:15:18 | INFO | Using BARX1 motif with MIXL1 model weights\n",
      "2021-07-19 15:15:18 | INFO | Using VAX1 motif with PDX1 model weights\n",
      "2021-07-19 15:15:19 | INFO | Using GSX2 motif with PDX1 model weights\n",
      "2021-07-19 15:15:19 | INFO | Using MEOX1 motif with PDX1 model weights\n",
      "2021-07-19 15:15:19 | INFO | Using PDX1 model\n",
      "2021-07-19 15:15:20 | INFO | Using HOXA2 motif with PDX1 model weights\n",
      "2021-07-19 15:15:20 | INFO | Using GATA6 motif with GATA4 model weights\n",
      "2021-07-19 15:15:21 | INFO | Using GATA4 model\n",
      "2021-07-19 15:15:21 | INFO | Using GATA5 motif with GATA4 model weights\n",
      "2021-07-19 15:15:21 | INFO | Using YY1 model\n",
      "2021-07-19 15:15:22 | INFO | Using NFATC3 model\n",
      "2021-07-19 15:15:22 | INFO | Using NFATC4 motif with NFATC3 model weights\n",
      "2021-07-19 15:15:23 | INFO | Using NFATC2 motif with NFATC3 model weights\n",
      "2021-07-19 15:15:23 | INFO | Using NFATC1 model\n",
      "2021-07-19 15:15:24 | INFO | Using TFEB motif with TFE3 model weights\n",
      "2021-07-19 15:15:24 | INFO | Using FOXK1 motif with FOXP1 model weights\n",
      "2021-07-19 15:15:25 | INFO | Using FOXO6 motif with FOXO1 model weights\n",
      "2021-07-19 15:15:25 | INFO | Using FOXJ2 model\n",
      "2021-07-19 15:15:25 | INFO | Using FOXL1 motif with FOXJ2 model weights\n",
      "2021-07-19 15:15:26 | INFO | Using FOXJ3 motif with FOXJ2 model weights\n",
      "2021-07-19 15:15:26 | INFO | Using FOXG1 motif with FOXP1 model weights\n",
      "2021-07-19 15:15:27 | INFO | Using FOXO4 motif with FOXO1 model weights\n",
      "2021-07-19 15:15:27 | INFO | Using FOXO1 model\n",
      "2021-07-19 15:15:27 | INFO | Using FOXP1 model\n",
      "2021-07-19 15:15:28 | INFO | Using FOXN2 motif with FOXJ2 model weights\n",
      "2021-07-19 15:15:28 | INFO | Using FOXN3 motif with FOXJ2 model weights\n",
      "2021-07-19 15:15:29 | INFO | Using FOXJ1 motif with FOXO1 model weights\n",
      "2021-07-19 15:15:29 | INFO | Using TFDP1 model\n",
      "2021-07-19 15:15:29 | INFO | Using E2F7 model\n",
      "2021-07-19 15:15:30 | INFO | Using XBP1 motif with CREB3 model weights\n",
      "2021-07-19 15:15:30 | INFO | Using NKX2-5 motif with GATA2 model weights\n",
      "2021-07-19 15:15:31 | INFO | Using NKX3-1 motif with GATA2 model weights\n",
      "2021-07-19 15:15:31 | INFO | Using NKX2-6 motif with GATA2 model weights\n",
      "2021-07-19 15:15:31 | INFO | Using NKX3-2 motif with GATA2 model weights\n",
      "2021-07-19 15:15:32 | INFO | Using NKX2-3 motif with GATA2 model weights\n",
      "2021-07-19 15:15:32 | INFO | Using NKX2-4 motif with GATA2 model weights\n",
      "2021-07-19 15:15:33 | INFO | Using FOXF1 motif with FOXK2 model weights\n",
      "2021-07-19 15:15:33 | INFO | Using HOXB3 motif with MIXL1 model weights\n",
      "2021-07-19 15:15:33 | INFO | Using VENTX motif with MIXL1 model weights\n",
      "2021-07-19 15:15:34 | INFO | Using MIXL1 model\n",
      "2021-07-19 15:15:34 | INFO | Using ZBTB7B model\n",
      "2021-07-19 15:15:34 | INFO | Using MSGN1 motif with NEUROD1 model weights\n",
      "2021-07-19 15:15:35 | INFO | Using BHLHE23 motif with NEUROD1 model weights\n",
      "2021-07-19 15:15:35 | INFO | Using NFIC model\n",
      "2021-07-19 15:15:36 | INFO | Using NFIA model\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:15:36 | INFO | Using NFIX motif with NFIB model weights\n",
      "2021-07-19 15:15:36 | INFO | Using NFIB model\n",
      "2021-07-19 15:15:37 | INFO | Using NFYA model\n",
      "2021-07-19 15:15:37 | INFO | Using FOXO3 model\n",
      "2021-07-19 15:15:37 | INFO | Using MLX model\n",
      "2021-07-19 15:15:38 | INFO | Using TCF12 model\n",
      "2021-07-19 15:15:38 | INFO | Using SCRT2 motif with ZEB1 model weights\n",
      "2021-07-19 15:15:39 | INFO | Using ATF1 model\n",
      "2021-07-19 15:15:39 | INFO | Using CREB1 model\n",
      "2021-07-19 15:15:39 | INFO | Using ZNF274 model\n",
      "2021-07-19 15:15:40 | INFO | Using CTCF model\n",
      "2021-07-19 15:15:40 | INFO | Using CTCFL model\n",
      "2021-07-19 15:15:40 | INFO | Using ZBTB33 model\n",
      "2021-07-19 15:15:41 | INFO | Using LHX6 motif with PDX1 model weights\n",
      "2021-07-19 15:15:41 | INFO | Using ONECUT3 motif with CUX1 model weights\n",
      "2021-07-19 15:15:42 | INFO | Using ONECUT2 motif with CUX1 model weights\n",
      "2021-07-19 15:15:42 | INFO | Using ONECUT1 motif with CUX1 model weights\n",
      "2021-07-19 15:15:42 | INFO | Using DUX4 motif with PDX1 model weights\n",
      "2021-07-19 15:15:43 | INFO | No match for BBX based on motifs\n",
      "2021-07-19 15:15:43 | INFO | No related TF found for BBX, using general model\n",
      "2021-07-19 15:15:43 | INFO | Using HOXD8 motif with PDX1 model weights\n",
      "2021-07-19 15:15:44 | INFO | Using KLF1 model\n",
      "2021-07-19 15:15:44 | INFO | Using SALL4 motif with KLF1 model weights\n",
      "2021-07-19 15:15:44 | INFO | Using MEF2C model\n",
      "2021-07-19 15:15:45 | INFO | Using MEF2D model\n",
      "2021-07-19 15:15:45 | INFO | Using MEF2B model\n",
      "2021-07-19 15:15:45 | INFO | Using MEF2A model\n",
      "2021-07-19 15:15:46 | INFO | Using EBF1 model\n",
      "2021-07-19 15:15:46 | INFO | Using EBF3 motif with EBF1 model weights\n",
      "2021-07-19 15:15:47 | INFO | Using LYL1 motif with TAL1 model weights\n",
      "2021-07-19 15:15:47 | INFO | Using PAX6 motif with PDX1 model weights\n",
      "2021-07-19 15:15:47 | INFO | Using MYOD1 motif with TCF3 model weights\n",
      "2021-07-19 15:15:48 | INFO | Using ASCL1 motif with TCF12 model weights\n",
      "2021-07-19 15:15:48 | INFO | Using MYOG motif with TCF12 model weights\n",
      "2021-07-19 15:15:48 | INFO | No match for ZNF35 based on motifs\n",
      "2021-07-19 15:15:48 | INFO | No related TF found for ZNF35, using general model\n",
      "2021-07-19 15:15:49 | INFO | Using FERD3L motif with NEUROD1 model weights\n",
      "2021-07-19 15:15:49 | INFO | Using EMX1 motif with PDX1 model weights\n",
      "2021-07-19 15:15:50 | INFO | Using ZFP42 motif with USF1 model weights\n",
      "2021-07-19 15:15:50 | INFO | Using NR3C1 model\n",
      "2021-07-19 15:15:50 | INFO | Using PGR model\n",
      "2021-07-19 15:15:51 | INFO | Using AR model\n",
      "2021-07-19 15:15:51 | INFO | Using NR3C2 motif with PGR model weights\n",
      "2021-07-19 15:15:52 | INFO | Using SOX1 motif with SOX2 model weights\n",
      "2021-07-19 15:15:52 | INFO | Using SOX21 motif with SOX2 model weights\n",
      "2021-07-19 15:15:53 | INFO | Using SOX14 motif with SOX2 model weights\n",
      "2021-07-19 15:15:53 | INFO | No match for OSR2 based on motifs\n",
      "2021-07-19 15:15:53 | INFO | No related TF found for OSR2, using general model\n",
      "2021-07-19 15:15:53 | INFO | No match for OSR1 based on motifs\n",
      "2021-07-19 15:15:53 | INFO | No related TF found for OSR1, using general model\n",
      "2021-07-19 15:15:54 | INFO | Using SREBF2 model\n",
      "2021-07-19 15:15:54 | INFO | Using POU2F2 model\n",
      "2021-07-19 15:15:55 | INFO | Using POU2F3 motif with POU2F2 model weights\n",
      "2021-07-19 15:15:55 | INFO | Using POU5F1B motif with POU2F2 model weights\n",
      "2021-07-19 15:15:55 | INFO | Using REST model\n",
      "2021-07-19 15:15:56 | INFO | Using ZNF143 model\n",
      "2021-07-19 15:15:56 | INFO | Using ZNF76 motif with ZNF143 model weights\n",
      "2021-07-19 15:15:57 | INFO | Using NRF1 model\n",
      "2021-07-19 15:15:57 | INFO | Using ZNF711 motif with ZFX model weights\n",
      "2021-07-19 15:15:57 | INFO | Using ZFX model\n",
      "2021-07-19 15:15:58 | INFO | Using ZFY motif with ZFX model weights\n",
      "2021-07-19 15:15:58 | INFO | Using RELA model\n",
      "2021-07-19 15:15:59 | INFO | Using REL motif with RELA model weights\n",
      "2021-07-19 15:15:59 | INFO | Using ISL1 motif with PDX1 model weights\n",
      "2021-07-19 15:15:59 | INFO | Using HLX motif with MIXL1 model weights\n",
      "2021-07-19 15:16:00 | INFO | Using GSX1 motif with PDX1 model weights\n",
      "2021-07-19 15:16:00 | INFO | Using HOXC4 motif with PDX1 model weights\n",
      "2021-07-19 15:16:01 | INFO | Using HOXA3 motif with PDX1 model weights\n",
      "2021-07-19 15:16:01 | INFO | Using HOXA4 motif with PDX1 model weights\n",
      "2021-07-19 15:16:01 | INFO | Using VAX2 motif with PDX1 model weights\n",
      "2021-07-19 15:16:02 | INFO | Using RXRB model\n",
      "2021-07-19 15:16:02 | INFO | Using ESR1 model\n",
      "2021-07-19 15:16:03 | INFO | Using THRA model\n",
      "2021-07-19 15:16:03 | INFO | Using THRB model\n",
      "2021-07-19 15:16:03 | INFO | Using HNF1B motif with HNF1A model weights\n",
      "2021-07-19 15:16:04 | INFO | Using HNF1A model\n",
      "2021-07-19 15:16:04 | INFO | Using TFAP2A model\n",
      "2021-07-19 15:16:04 | INFO | Using TFAP2C model\n",
      "2021-07-19 15:16:05 | INFO | Using TFAP2E motif with TFAP2C model weights\n",
      "2021-07-19 15:16:05 | INFO | Using TFDP2 motif with TFDP1 model weights\n",
      "2021-07-19 15:16:06 | INFO | Using TEF motif with NFIL3 model weights\n",
      "2021-07-19 15:16:06 | INFO | No match for MYPOP based on motifs\n",
      "2021-07-19 15:16:06 | INFO | No related TF found for MYPOP, using general model\n",
      "2021-07-19 15:16:07 | INFO | Using JUN model\n",
      "2021-07-19 15:16:07 | INFO | Using BATF model\n",
      "2021-07-19 15:16:07 | INFO | Using IRF4 model\n",
      "2021-07-19 15:16:08 | INFO | Using SEBOX motif with PDX1 model weights\n",
      "2021-07-19 15:16:08 | INFO | Using SIM2 motif with AHR model weights\n",
      "2021-07-19 15:16:08 | INFO | Using CEBPE motif with NFIL3 model weights\n",
      "2021-07-19 15:16:09 | INFO | Using CEBPA model\n",
      "2021-07-19 15:16:09 | INFO | Using SP4 motif with SP5 model weights\n",
      "2021-07-19 15:16:10 | INFO | Using KLF8 motif with KLF5 model weights\n",
      "2021-07-19 15:16:10 | INFO | Using NR2F6 model\n",
      "2021-07-19 15:16:10 | INFO | Using PPARG model\n",
      "2021-07-19 15:16:11 | INFO | Using NR2F1 model\n",
      "2021-07-19 15:16:11 | INFO | Using NR1H2 motif with RXRB model weights\n",
      "2021-07-19 15:16:12 | INFO | Using PPARA motif with PPARG model weights\n",
      "2021-07-19 15:16:12 | INFO | Using PPARD motif with PPARG model weights\n",
      "2021-07-19 15:16:12 | INFO | Using SPIB motif with SPI1 model weights\n",
      "2021-07-19 15:16:13 | INFO | Using E2F6 model\n",
      "2021-07-19 15:16:13 | INFO | Using RBPJ model\n",
      "2021-07-19 15:16:14 | INFO | Using MYB model\n",
      "2021-07-19 15:16:14 | INFO | No match for IRX5 based on motifs\n",
      "2021-07-19 15:16:14 | INFO | No related TF found for IRX5, using general model\n",
      "2021-07-19 15:16:14 | INFO | No match for IRX2 based on motifs\n",
      "2021-07-19 15:16:14 | INFO | No related TF found for IRX2, using general model\n",
      "2021-07-19 15:16:15 | INFO | Using EGR3 motif with EGR1 model weights\n",
      "2021-07-19 15:16:15 | INFO | Using TBX4 motif with MGA model weights\n",
      "2021-07-19 15:16:16 | INFO | Using TBX5 motif with MGA model weights\n",
      "2021-07-19 15:16:16 | INFO | Using STAT6 motif with ELK1 model weights\n",
      "2021-07-19 15:16:16 | INFO | Using HAND1 motif with TCF3 model weights\n",
      "2021-07-19 15:16:17 | INFO | Using FOXB1 motif with FOXM1 model weights\n",
      "2021-07-19 15:16:17 | INFO | Using FOXC1 motif with FOXK2 model weights\n",
      "2021-07-19 15:16:17 | INFO | Using FOXC2 motif with FOXK2 model weights\n",
      "2021-07-19 15:16:18 | INFO | Using FOSL2 model\n",
      "2021-07-19 15:16:18 | INFO | Using TCF7L2 model\n",
      "2021-07-19 15:16:19 | INFO | Using PRDM1 model\n",
      "2021-07-19 15:16:19 | INFO | Using ZNF263 model\n",
      "2021-07-19 15:16:20 | INFO | Using TCF4 motif with TCF3 model weights\n",
      "2021-07-19 15:16:20 | INFO | Using ZNF740 model\n",
      "2021-07-19 15:16:20 | INFO | Using POU6F1 motif with PDX1 model weights\n",
      "2021-07-19 15:16:21 | INFO | Using EMX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:16:21 | INFO | Using NOTO motif with PDX1 model weights\n",
      "2021-07-19 15:16:21 | INFO | Using MNX1 motif with PDX1 model weights\n",
      "2021-07-19 15:16:22 | INFO | Using HOXA7 motif with PDX1 model weights\n",
      "2021-07-19 15:16:22 | INFO | Using FOXP2 motif with FOXO3 model weights\n",
      "2021-07-19 15:16:23 | INFO | Using FOXP4 motif with FOXO3 model weights\n",
      "2021-07-19 15:16:23 | INFO | Using FOXP3 motif with FOXO3 model weights\n",
      "2021-07-19 15:16:24 | INFO | Using SOX10 motif with SOX2 model weights\n",
      "2021-07-19 15:16:24 | INFO | Using EVX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:16:24 | INFO | Using NKX1-1 motif with MIXL1 model weights\n",
      "2021-07-19 15:16:25 | INFO | Using NKX1-2 motif with MIXL1 model weights\n",
      "2021-07-19 15:16:25 | INFO | Using EVX1 motif with MIXL1 model weights\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:16:26 | INFO | Using BARX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:16:26 | INFO | Using PBX1 motif with PBX2 model weights\n",
      "2021-07-19 15:16:26 | INFO | Using ARNTL motif with BHLHE40 model weights\n",
      "2021-07-19 15:16:27 | INFO | Using HOXC12 motif with PDX1 model weights\n",
      "2021-07-19 15:16:27 | INFO | Using HOXA11 motif with PDX1 model weights\n",
      "2021-07-19 15:16:28 | INFO | Using HOXA9 motif with MEIS2 model weights\n",
      "2021-07-19 15:16:28 | INFO | Using HOXD11 motif with PDX1 model weights\n",
      "2021-07-19 15:16:28 | INFO | Using HOXD9 motif with PDX1 model weights\n",
      "2021-07-19 15:16:29 | INFO | Using HOXC11 motif with PDX1 model weights\n",
      "2021-07-19 15:16:29 | INFO | Using HOXD12 motif with PDX1 model weights\n",
      "2021-07-19 15:16:30 | INFO | Using HOXC10 motif with PDX1 model weights\n",
      "2021-07-19 15:16:30 | INFO | Using HOXA10 motif with PDX1 model weights\n",
      "2021-07-19 15:16:30 | INFO | Using STAT1 model\n",
      "2021-07-19 15:16:31 | INFO | Using STAT4 motif with STAT5A model weights\n",
      "2021-07-19 15:16:31 | INFO | Using STAT3 model\n",
      "2021-07-19 15:16:31 | INFO | Using STAT5A model\n",
      "2021-07-19 15:16:32 | INFO | Using NKX6-3 motif with PDX1 model weights\n",
      "2021-07-19 15:16:32 | INFO | Using NKX6-2 motif with PDX1 model weights\n",
      "2021-07-19 15:16:33 | INFO | Using LHX8 motif with PDX1 model weights\n",
      "2021-07-19 15:16:33 | INFO | Using THAP11 model\n",
      "2021-07-19 15:16:33 | INFO | Using MAFB motif with MAFG model weights\n",
      "2021-07-19 15:16:34 | INFO | Using MAFG model\n",
      "2021-07-19 15:16:34 | INFO | Using MAFF model\n",
      "2021-07-19 15:16:35 | INFO | Using MAFA motif with MAFG model weights\n",
      "2021-07-19 15:16:35 | INFO | No match for ZSCAN10 based on motifs\n",
      "2021-07-19 15:16:35 | INFO | No related TF found for ZSCAN10, using general model\n",
      "2021-07-19 15:16:35 | INFO | No match for MZF1 based on motifs\n",
      "2021-07-19 15:16:35 | INFO | No related TF found for MZF1, using general model\n",
      "2021-07-19 15:16:36 | INFO | No match for HIC1 based on motifs\n",
      "2021-07-19 15:16:36 | INFO | No related TF found for HIC1, using general model\n",
      "2021-07-19 15:16:36 | INFO | Using LEF1 model\n",
      "2021-07-19 15:16:36 | INFO | Using TCF7L1 motif with LEF1 model weights\n",
      "2021-07-19 15:16:37 | INFO | Using FOXA2 model\n",
      "2021-07-19 15:16:37 | INFO | Using ELF2 motif with ETV1 model weights\n",
      "2021-07-19 15:16:38 | INFO | No match for ZBTB14 based on motifs\n",
      "2021-07-19 15:16:38 | INFO | No related TF found for ZBTB14, using general model\n",
      "2021-07-19 15:16:38 | INFO | Using ESR2 motif with ESR1 model weights\n",
      "2021-07-19 15:16:38 | INFO | No match for ZNF333 based on motifs\n",
      "2021-07-19 15:16:38 | INFO | No related TF found for ZNF333, using general model\n",
      "2021-07-19 15:16:39 | INFO | Using AHRR motif with AHR model weights\n",
      "2021-07-19 15:16:39 | INFO | Using ATF2 model\n",
      "2021-07-19 15:16:40 | INFO | Using CREM model\n",
      "2021-07-19 15:16:40 | INFO | Using CREB5 motif with ATF2 model weights\n",
      "2021-07-19 15:16:40 | INFO | Using FOXM1 model\n",
      "2021-07-19 15:16:41 | INFO | No match for GLIS1 based on motifs\n",
      "2021-07-19 15:16:41 | INFO | No related TF found for GLIS1, using general model\n",
      "2021-07-19 15:16:41 | INFO | No match for GLIS2 based on motifs\n",
      "2021-07-19 15:16:41 | INFO | No related TF found for GLIS2, using general model\n",
      "2021-07-19 15:16:41 | INFO | Using BHLHE40 model\n",
      "2021-07-19 15:16:42 | INFO | Using HES7 motif with HES1 model weights\n",
      "2021-07-19 15:16:42 | INFO | Using HES5 motif with HES1 model weights\n",
      "2021-07-19 15:16:43 | INFO | Using HES1 model\n",
      "2021-07-19 15:16:43 | INFO | Using HES4 motif with HES1 model weights\n",
      "2021-07-19 15:16:43 | INFO | Using HES2 motif with HES1 model weights\n",
      "2021-07-19 15:16:44 | INFO | Using BCL11A model\n",
      "2021-07-19 15:16:44 | INFO | Using E4F1 model\n",
      "2021-07-19 15:16:45 | INFO | Using ZNF148 model\n",
      "2021-07-19 15:16:45 | INFO | Using RARB motif with RARA model weights\n",
      "2021-07-19 15:16:45 | INFO | Using NR1I2 motif with HNF4A model weights\n",
      "2021-07-19 15:16:46 | INFO | Using VDR motif with NR2C1 model weights\n",
      "2021-07-19 15:16:46 | INFO | No match for DMRTA2 based on motifs\n",
      "2021-07-19 15:16:46 | INFO | No related TF found for DMRTA2, using general model\n",
      "2021-07-19 15:16:46 | INFO | Using NR4A1 model\n",
      "2021-07-19 15:16:47 | INFO | Using NR5A1 motif with NR5A2 model weights\n",
      "2021-07-19 15:16:47 | INFO | Using NR4A3 motif with NR4A1 model weights\n",
      "2021-07-19 15:16:48 | INFO | Using MSC motif with NEUROD1 model weights\n",
      "2021-07-19 15:16:48 | INFO | Using NR2F2 model\n",
      "2021-07-19 15:16:48 | INFO | Using HBP1 model\n",
      "2021-07-19 15:16:49 | INFO | No match for HEY1 based on motifs\n",
      "2021-07-19 15:16:49 | INFO | No related TF found for HEY1, using general model\n",
      "2021-07-19 15:16:49 | INFO | No match for HEY2 based on motifs\n",
      "2021-07-19 15:16:49 | INFO | No related TF found for HEY2, using general model\n",
      "2021-07-19 15:16:50 | INFO | No match for HEYL based on motifs\n",
      "2021-07-19 15:16:50 | INFO | No related TF found for HEYL, using general model\n",
      "2021-07-19 15:16:50 | INFO | Using MEIS1 motif with MEIS2 model weights\n",
      "2021-07-19 15:16:50 | INFO | Using MEIS3 motif with TGIF2 model weights\n",
      "2021-07-19 15:16:51 | INFO | Using MEIS2 model\n",
      "2021-07-19 15:16:51 | INFO | Using PKNOX1 model\n",
      "2021-07-19 15:16:52 | INFO | Using PKNOX2 motif with TGIF2 model weights\n",
      "2021-07-19 15:16:52 | INFO | Using TGIF2 model\n",
      "2021-07-19 15:16:53 | INFO | Using TGIF2LY motif with TGIF2 model weights\n",
      "2021-07-19 15:16:53 | INFO | Using MLXIPL motif with MLX model weights\n",
      "2021-07-19 15:16:53 | INFO | Using MLXIP motif with MLX model weights\n",
      "2021-07-19 15:16:54 | INFO | Using BHLHE41 motif with MLX model weights\n",
      "2021-07-19 15:16:54 | INFO | Using ARNTL2 motif with BHLHE40 model weights\n",
      "2021-07-19 15:16:55 | INFO | Using CDX2 motif with PDX1 model weights\n",
      "2021-07-19 15:16:55 | INFO | Using ESRRA model\n",
      "2021-07-19 15:16:55 | INFO | No match for ZNF436 based on motifs\n",
      "2021-07-19 15:16:55 | INFO | No related TF found for ZNF436, using general model\n",
      "2021-07-19 15:16:56 | INFO | Using ATOH8 motif with TAL1 model weights\n",
      "2021-07-19 15:16:56 | INFO | No match for DMRTC2 based on motifs\n",
      "2021-07-19 15:16:56 | INFO | No related TF found for DMRTC2, using general model\n",
      "2021-07-19 15:16:57 | INFO | Using HOMEZ model\n",
      "2021-07-19 15:16:57 | INFO | Using TP63 motif with TP73 model weights\n",
      "2021-07-19 15:16:57 | INFO | Using TP73 model\n",
      "2021-07-19 15:16:58 | INFO | Using TP53 model\n",
      "2021-07-19 15:16:58 | INFO | Using PRDM14 model\n",
      "2021-07-19 15:16:59 | INFO | Using SPIC motif with SPI1 model weights\n",
      "2021-07-19 15:16:59 | INFO | No match for ZNF382 based on motifs\n",
      "2021-07-19 15:16:59 | INFO | No related TF found for ZNF382, using general model\n",
      "2021-07-19 15:16:59 | INFO | Using TFAP2B motif with TFAP2C model weights\n",
      "2021-07-19 15:17:00 | INFO | Using RFX1 model\n",
      "2021-07-19 15:17:00 | INFO | Using RFX2 motif with RFX3 model weights\n",
      "2021-07-19 15:17:00 | INFO | Using BCL6 model\n",
      "2021-07-19 15:17:01 | INFO | No match for TFCP2L1 based on motifs\n",
      "2021-07-19 15:17:01 | INFO | No related TF found for TFCP2L1, using general model\n",
      "2021-07-19 15:17:01 | INFO | No match for UBP1 based on motifs\n",
      "2021-07-19 15:17:01 | INFO | No related TF found for UBP1, using general model\n",
      "2021-07-19 15:17:02 | INFO | Using TLX2 motif with NFIB model weights\n",
      "2021-07-19 15:17:02 | INFO | Using TLX1 motif with NFIB model weights\n",
      "2021-07-19 15:17:02 | INFO | Using TLX3 motif with NFIB model weights\n",
      "2021-07-19 15:17:03 | INFO | Using ISL2 motif with PDX1 model weights\n",
      "2021-07-19 15:17:03 | INFO | Using BATF3 model\n",
      "2021-07-19 15:17:04 | INFO | Using FOXD1 motif with FOXO1 model weights\n",
      "2021-07-19 15:17:04 | INFO | Using PROP1 motif with MIXL1 model weights\n",
      "2021-07-19 15:17:04 | INFO | Using PHOX2A motif with MIXL1 model weights\n",
      "2021-07-19 15:17:05 | INFO | Using PAX7 motif with PDX1 model weights\n",
      "2021-07-19 15:17:05 | INFO | Using PHOX2B motif with MIXL1 model weights\n",
      "2021-07-19 15:17:06 | INFO | Using IKZF1 model\n",
      "2021-07-19 15:17:06 | INFO | Using IKZF4 motif with IKZF1 model weights\n",
      "2021-07-19 15:17:06 | INFO | No match for MTF1 based on motifs\n",
      "2021-07-19 15:17:06 | INFO | No related TF found for MTF1, using general model\n",
      "2021-07-19 15:17:07 | INFO | Using TCF7 model\n",
      "2021-07-19 15:17:07 | INFO | Using OVOL2 motif with KLF9 model weights\n",
      "2021-07-19 15:17:07 | INFO | Using KLF9 model\n",
      "2021-07-19 15:17:08 | INFO | Using KLF13 model\n",
      "2021-07-19 15:17:08 | INFO | Using OVOL1 motif with KLF9 model weights\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:17:09 | INFO | Using KLF14 motif with KLF9 model weights\n",
      "2021-07-19 15:17:09 | INFO | Using TBX2 motif with MGA model weights\n",
      "2021-07-19 15:17:09 | INFO | Using TBX15 motif with MGA model weights\n",
      "2021-07-19 15:17:10 | INFO | Using MGA model\n",
      "2021-07-19 15:17:10 | INFO | Using TBX22 motif with MGA model weights\n",
      "2021-07-19 15:17:11 | INFO | Using TBX18 motif with MGA model weights\n",
      "2021-07-19 15:17:11 | INFO | Using ETV5 model\n",
      "2021-07-19 15:17:11 | INFO | Using DBX2 motif with PDX1 model weights\n",
      "2021-07-19 15:17:12 | INFO | Using BARHL1 motif with MIXL1 model weights\n",
      "2021-07-19 15:17:12 | INFO | Using BARHL2 motif with MIXL1 model weights\n",
      "2021-07-19 15:17:12 | INFO | Using TBX19 motif with MGA model weights\n",
      "2021-07-19 15:17:13 | INFO | No match for HMX3 based on motifs\n",
      "2021-07-19 15:17:13 | INFO | No related TF found for HMX3, using general model\n",
      "2021-07-19 15:17:13 | INFO | No match for HMX1 based on motifs\n",
      "2021-07-19 15:17:13 | INFO | No related TF found for HMX1, using general model\n",
      "2021-07-19 15:17:14 | INFO | Using NKX2-8 motif with GATA2 model weights\n",
      "2021-07-19 15:17:14 | INFO | Using NKX2-2 motif with GATA2 model weights\n",
      "2021-07-19 15:17:14 | INFO | Using NKX2-1 motif with GATA2 model weights\n",
      "2021-07-19 15:17:15 | INFO | Using YY2 motif with YY1 model weights\n",
      "2021-07-19 15:17:15 | INFO | Using CDX1 motif with PDX1 model weights\n",
      "2021-07-19 15:17:15 | INFO | Using NKX6-1 motif with PDX1 model weights\n",
      "2021-07-19 15:17:16 | INFO | Using HOXC6 motif with PDX1 model weights\n",
      "2021-07-19 15:17:16 | INFO | Using ZIC3 motif with ZBTB7B model weights\n",
      "2021-07-19 15:17:17 | INFO | Using ZIC2 motif with ZBTB7B model weights\n",
      "2021-07-19 15:17:17 | INFO | Using TFEC motif with TFE3 model weights\n",
      "2021-07-19 15:17:18 | INFO | Using IRF6 motif with IRF5 model weights\n",
      "2021-07-19 15:17:18 | INFO | No match for HIVEP1 based on motifs\n",
      "2021-07-19 15:17:18 | INFO | No related TF found for HIVEP1, using general model\n",
      "2021-07-19 15:17:18 | INFO | No match for HIVEP3 based on motifs\n",
      "2021-07-19 15:17:18 | INFO | No related TF found for HIVEP3, using general model\n",
      "2021-07-19 15:17:19 | INFO | No match for HIVEP2 based on motifs\n",
      "2021-07-19 15:17:19 | INFO | No related TF found for HIVEP2, using general model\n",
      "2021-07-19 15:17:19 | INFO | Using MNT model\n",
      "2021-07-19 15:17:19 | INFO | Using ETV6 model\n",
      "2021-07-19 15:17:20 | INFO | Using GRHL1 motif with GRHL2 model weights\n",
      "2021-07-19 15:17:20 | INFO | Using GRHL2 model\n",
      "2021-07-19 15:17:21 | INFO | Using ZNF683 motif with PRDM1 model weights\n",
      "2021-07-19 15:17:21 | INFO | Using BHLHE22 motif with NEUROD1 model weights\n",
      "2021-07-19 15:17:21 | INFO | Using OLIG3 motif with NEUROD1 model weights\n",
      "2021-07-19 15:17:22 | INFO | Using OLIG1 motif with NEUROD1 model weights\n",
      "2021-07-19 15:17:22 | INFO | No match for HIC2 based on motifs\n",
      "2021-07-19 15:17:22 | INFO | No related TF found for HIC2, using general model\n",
      "2021-07-19 15:17:22 | INFO | Using NR5A2 model\n",
      "2021-07-19 15:17:23 | INFO | Using TFAP4 model\n",
      "2021-07-19 15:17:23 | INFO | Using MESP1 motif with TCF3 model weights\n",
      "2021-07-19 15:17:24 | INFO | Using MESP2 motif with TCF3 model weights\n",
      "2021-07-19 15:17:24 | INFO | Using KLF16 model\n",
      "2021-07-19 15:17:24 | INFO | Using KLF11 motif with KLF9 model weights\n",
      "2021-07-19 15:17:25 | INFO | Using STAT5B motif with STAT5A model weights\n",
      "2021-07-19 15:17:25 | INFO | Using ZBTB7C motif with ZBTB7B model weights\n",
      "2021-07-19 15:17:25 | INFO | Using GLI2 motif with ZBTB7B model weights\n",
      "2021-07-19 15:17:26 | INFO | Using GLI1 motif with ZBTB7B model weights\n",
      "2021-07-19 15:17:26 | INFO | Using ATF7 model\n",
      "2021-07-19 15:17:27 | INFO | Using ZNF691 motif with ZNF263 model weights\n",
      "2021-07-19 15:17:27 | INFO | Using HINFP model\n",
      "2021-07-19 15:17:27 | INFO | Using TEAD1 model\n",
      "2021-07-19 15:17:28 | INFO | Using TEAD4 model\n",
      "2021-07-19 15:17:28 | INFO | Using TEAD3 model\n",
      "2021-07-19 15:17:29 | INFO | Using TEAD2 model\n",
      "2021-07-19 15:17:29 | INFO | Using HSF2 motif with HSF1 model weights\n",
      "2021-07-19 15:17:29 | INFO | Using NRL motif with MAFG model weights\n",
      "2021-07-19 15:17:30 | INFO | Using ZNF281 motif with ZNF148 model weights\n",
      "2021-07-19 15:17:30 | INFO | Using TWIST1 motif with TCF3 model weights\n",
      "2021-07-19 15:17:31 | INFO | Using TAL2 motif with TAL1 model weights\n",
      "2021-07-19 15:17:31 | INFO | Using IRF9 model\n",
      "2021-07-19 15:17:31 | INFO | Using IRF8 motif with IRF4 model weights\n",
      "2021-07-19 15:17:32 | INFO | Using BCL11B motif with IRF4 model weights\n",
      "2021-07-19 15:17:32 | INFO | Using ZNF75A model\n",
      "2021-07-19 15:17:33 | INFO | Using ZNF75D motif with ZNF75A model weights\n",
      "2021-07-19 15:17:33 | INFO | No match for HMX2 based on motifs\n",
      "2021-07-19 15:17:33 | INFO | No related TF found for HMX2, using general model\n",
      "2021-07-19 15:17:33 | INFO | Using STAT2 model\n",
      "2021-07-19 15:17:34 | INFO | Using NFYB model\n",
      "2021-07-19 15:17:34 | INFO | Using NFYC model\n",
      "2021-07-19 15:17:35 | INFO | Using ETV7 motif with MXI1 model weights\n",
      "2021-07-19 15:17:35 | INFO | Using KLF15 motif with MAZ model weights\n",
      "2021-07-19 15:17:35 | INFO | Using MAZ model\n",
      "2021-07-19 15:17:36 | INFO | Using ZNF467 motif with MAZ model weights\n",
      "2021-07-19 15:17:36 | INFO | Using WT1 motif with SP1 model weights\n",
      "2021-07-19 15:17:36 | INFO | Using MEOX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:17:37 | INFO | Using LHX2 motif with PDX1 model weights\n",
      "2021-07-19 15:17:37 | INFO | Using GFI1 motif with GFI1B model weights\n",
      "2021-07-19 15:17:38 | INFO | Using GFI1B model\n",
      "2021-07-19 15:17:38 | INFO | Using ESRRG motif with ESRRA model weights\n",
      "2021-07-19 15:17:38 | INFO | Using TFCP2 motif with GRHL2 model weights\n",
      "2021-07-19 15:17:39 | INFO | No match for ZSCAN16 based on motifs\n",
      "2021-07-19 15:17:39 | INFO | No related TF found for ZSCAN16, using general model\n",
      "2021-07-19 15:17:39 | INFO | No match for ZNF586 based on motifs\n",
      "2021-07-19 15:17:39 | INFO | No related TF found for ZNF586, using general model\n",
      "2021-07-19 15:17:40 | INFO | Using NR1I3 motif with NR2C1 model weights\n",
      "2021-07-19 15:17:40 | INFO | Using KLF10 model\n",
      "2021-07-19 15:17:40 | INFO | Using ZBED1 model\n",
      "2021-07-19 15:17:41 | INFO | No match for SPDEF based on motifs\n",
      "2021-07-19 15:17:41 | INFO | No related TF found for SPDEF, using general model\n",
      "2021-07-19 15:17:41 | INFO | No match for HSFY1 based on motifs\n",
      "2021-07-19 15:17:41 | INFO | No related TF found for HSFY1, using general model\n",
      "2021-07-19 15:17:41 | INFO | Using PAX3 motif with PDX1 model weights\n",
      "2021-07-19 15:17:42 | INFO | Using NFKB2 motif with RELA model weights\n",
      "2021-07-19 15:17:42 | INFO | Using SPZ1 motif with ZNF589 model weights\n",
      "2021-07-19 15:17:43 | INFO | No match for ZNF354C based on motifs\n",
      "2021-07-19 15:17:43 | INFO | No related TF found for ZNF354C, using general model\n",
      "2021-07-19 15:17:43 | INFO | Using TBX3 model\n",
      "2021-07-19 15:17:43 | INFO | Using CDC5L model\n",
      "2021-07-19 15:17:44 | INFO | No match for ZNF260 based on motifs\n",
      "2021-07-19 15:17:44 | INFO | No related TF found for ZNF260, using general model\n",
      "2021-07-19 15:17:44 | INFO | No match for ZNF146 based on motifs\n",
      "2021-07-19 15:17:44 | INFO | No related TF found for ZNF146, using general model\n",
      "2021-07-19 15:17:44 | INFO | No match for ZNF713 based on motifs\n",
      "2021-07-19 15:17:44 | INFO | No related TF found for ZNF713, using general model\n",
      "2021-07-19 15:17:45 | INFO | Using IKZF2 model\n",
      "2021-07-19 15:17:45 | INFO | No match for ZNF322 based on motifs\n",
      "2021-07-19 15:17:45 | INFO | No related TF found for ZNF322, using general model\n",
      "2021-07-19 15:17:46 | INFO | No match for ZBTB3 based on motifs\n",
      "2021-07-19 15:17:46 | INFO | No related TF found for ZBTB3, using general model\n",
      "2021-07-19 15:17:46 | INFO | No match for ZNF528 based on motifs\n",
      "2021-07-19 15:17:46 | INFO | No related TF found for ZNF528, using general model\n",
      "2021-07-19 15:17:46 | INFO | No match for FOXN1 based on motifs\n",
      "2021-07-19 15:17:46 | INFO | No related TF found for FOXN1, using general model\n",
      "2021-07-19 15:17:47 | INFO | Using PRDM16 motif with RUNX1 model weights\n",
      "2021-07-19 15:17:47 | INFO | Using ZNF8 model\n",
      "2021-07-19 15:17:48 | INFO | No match for ZNF667 based on motifs\n",
      "2021-07-19 15:17:48 | INFO | No related TF found for ZNF667, using general model\n",
      "2021-07-19 15:17:48 | INFO | No match for ZNF423 based on motifs\n",
      "2021-07-19 15:17:48 | INFO | No related TF found for ZNF423, using general model\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:17:49 | INFO | No match for ZNF354A based on motifs\n",
      "2021-07-19 15:17:49 | INFO | No related TF found for ZNF354A, using general model\n",
      "2021-07-19 15:17:49 | INFO | No match for ZFHX3 based on motifs\n",
      "2021-07-19 15:17:49 | INFO | No related TF found for ZFHX3, using general model\n",
      "2021-07-19 15:17:49 | INFO | No match for ZSCAN4 based on motifs\n",
      "2021-07-19 15:17:49 | INFO | No related TF found for ZSCAN4, using general model\n",
      "2021-07-19 15:17:50 | INFO | Using CUX2 motif with CUX1 model weights\n",
      "2021-07-19 15:17:50 | INFO | Using SIX5 model\n",
      "2021-07-19 15:17:50 | INFO | Using DBP motif with NFIL3 model weights\n",
      "2021-07-19 15:17:51 | INFO | No match for ZKSCAN3 based on motifs\n",
      "2021-07-19 15:17:51 | INFO | No related TF found for ZKSCAN3, using general model\n",
      "2021-07-19 15:17:51 | INFO | No match for ZKSCAN4 based on motifs\n",
      "2021-07-19 15:17:51 | INFO | No related TF found for ZKSCAN4, using general model\n",
      "2021-07-19 15:17:52 | INFO | No match for ZNF394 based on motifs\n",
      "2021-07-19 15:17:52 | INFO | No related TF found for ZNF394, using general model\n",
      "2021-07-19 15:17:52 | INFO | Using TWIST2 motif with TCF3 model weights\n",
      "2021-07-19 15:17:52 | INFO | Using TCFL5 motif with MYC model weights\n",
      "2021-07-19 15:17:53 | INFO | No match for ZNF554 based on motifs\n",
      "2021-07-19 15:17:53 | INFO | No related TF found for ZNF554, using general model\n",
      "2021-07-19 15:17:53 | INFO | Using ESRRB motif with ESRRA model weights\n",
      "2021-07-19 15:17:54 | INFO | Using DRGX motif with MIXL1 model weights\n",
      "2021-07-19 15:17:54 | INFO | Using SIX4 model\n",
      "2021-07-19 15:17:54 | INFO | Using E2F8 model\n",
      "2021-07-19 15:17:55 | INFO | No match for ZFP82 based on motifs\n",
      "2021-07-19 15:17:55 | INFO | No related TF found for ZFP82, using general model\n",
      "2021-07-19 15:17:55 | INFO | No match for ZBTB12 based on motifs\n",
      "2021-07-19 15:17:55 | INFO | No related TF found for ZBTB12, using general model\n",
      "2021-07-19 15:17:55 | INFO | Using TGIF2LX motif with TGIF2 model weights\n",
      "2021-07-19 15:17:56 | INFO | No match for ZNF219 based on motifs\n",
      "2021-07-19 15:17:56 | INFO | No related TF found for ZNF219, using general model\n",
      "2021-07-19 15:17:56 | INFO | Using ZNF589 model\n",
      "2021-07-19 15:17:57 | INFO | Using DUXA motif with PDX1 model weights\n",
      "2021-07-19 15:17:57 | INFO | No match for ZNF524 based on motifs\n",
      "2021-07-19 15:17:57 | INFO | No related TF found for ZNF524, using general model\n",
      "2021-07-19 15:17:57 | INFO | Using MECOM motif with RUNX1 model weights\n",
      "2021-07-19 15:17:58 | INFO | No match for FEZF1 based on motifs\n",
      "2021-07-19 15:17:58 | INFO | No related TF found for FEZF1, using general model\n",
      "2021-07-19 15:17:58 | INFO | No match for FEZF2 based on motifs\n",
      "2021-07-19 15:17:58 | INFO | No related TF found for FEZF2, using general model\n",
      "2021-07-19 15:17:58 | INFO | No match for ZNF350 based on motifs\n",
      "2021-07-19 15:17:58 | INFO | No related TF found for ZNF350, using general model\n",
      "2021-07-19 15:17:59 | INFO | No match for NFAT5 based on motifs\n",
      "2021-07-19 15:17:59 | INFO | No related TF found for NFAT5, using general model\n",
      "2021-07-19 15:17:59 | INFO | No match for ZNF264 based on motifs\n",
      "2021-07-19 15:17:59 | INFO | No related TF found for ZNF264, using general model\n",
      "2021-07-19 15:18:00 | INFO | Using CEBPD model\n",
      "2021-07-19 15:18:00 | INFO | Using CUX1 model\n",
      "2021-07-19 15:18:00 | INFO | Using CREB3L1 model\n",
      "2021-07-19 15:18:01 | INFO | Using CREB3L2 motif with CREB3L1 model weights\n",
      "2021-07-19 15:18:01 | INFO | Using CREB3L3 motif with CREB3L1 model weights\n",
      "2021-07-19 15:18:02 | INFO | Using ZBTB17 motif with SP1 model weights\n",
      "2021-07-19 15:18:02 | INFO | No match for RREB1 based on motifs\n",
      "2021-07-19 15:18:02 | INFO | No related TF found for RREB1, using general model\n",
      "2021-07-19 15:18:02 | INFO | Using ZNF384 model\n",
      "2021-07-19 15:18:03 | INFO | Using PAX2 motif with PAX8 model weights\n",
      "2021-07-19 15:18:03 | INFO | No match for DMRTB1 based on motifs\n",
      "2021-07-19 15:18:03 | INFO | No related TF found for DMRTB1, using general model\n",
      "2021-07-19 15:18:04 | INFO | No match for ZNF549 based on motifs\n",
      "2021-07-19 15:18:04 | INFO | No related TF found for ZNF549, using general model\n",
      "2021-07-19 15:18:04 | INFO | No match for ZNF232 based on motifs\n",
      "2021-07-19 15:18:04 | INFO | No related TF found for ZNF232, using general model\n",
      "2021-07-19 15:18:04 | INFO | Using GLI3 motif with ZBTB7B model weights\n",
      "2021-07-19 15:18:05 | INFO | Using SMAD1 model\n",
      "2021-07-19 15:18:05 | INFO | Using ZBTB16 model\n",
      "2021-07-19 15:18:05 | INFO | Using DBX1 motif with NFIB model weights\n",
      "2021-07-19 15:18:06 | INFO | No match for GZF1 based on motifs\n",
      "2021-07-19 15:18:06 | INFO | No related TF found for GZF1, using general model\n",
      "2021-07-19 15:18:06 | INFO | No match for SP110 based on motifs\n",
      "2021-07-19 15:18:06 | INFO | No related TF found for SP110, using general model\n",
      "2021-07-19 15:18:07 | INFO | Using ELF4 model\n",
      "2021-07-19 15:18:07 | INFO | Using LBX2 motif with MIXL1 model weights\n",
      "2021-07-19 15:18:07 | INFO | Using IRF7 motif with IRF9 model weights\n",
      "2021-07-19 15:18:08 | INFO | No match for ZNF85 based on motifs\n",
      "2021-07-19 15:18:08 | INFO | No related TF found for ZNF85, using general model\n",
      "2021-07-19 15:18:08 | INFO | Using CREB3 model\n",
      "2021-07-19 15:18:08 | INFO | Using CREB3L4 motif with CREB3 model weights\n",
      "2021-07-19 15:18:09 | INFO | Using RELB model\n",
      "2021-07-19 15:18:09 | INFO | No match for BCL6B based on motifs\n",
      "2021-07-19 15:18:09 | INFO | No related TF found for BCL6B, using general model\n",
      "2021-07-19 15:18:10 | INFO | Using PLAG1 model\n",
      "2021-07-19 15:18:10 | INFO | No match for ZBTB18 based on motifs\n",
      "2021-07-19 15:18:10 | INFO | No related TF found for ZBTB18, using general model\n",
      "2021-07-19 15:18:10 | INFO | No match for ZBTB42 based on motifs\n",
      "2021-07-19 15:18:10 | INFO | No related TF found for ZBTB42, using general model\n",
      "2021-07-19 15:18:11 | INFO | No match for ZNF563 based on motifs\n",
      "2021-07-19 15:18:11 | INFO | No related TF found for ZNF563, using general model\n",
      "2021-07-19 15:18:11 | INFO | Using CIC motif with MAFG model weights\n",
      "2021-07-19 15:18:11 | INFO | Using HAND2 motif with TCF3 model weights\n",
      "2021-07-19 15:18:12 | INFO | Using IRF5 model\n",
      "2021-07-19 15:18:12 | INFO | Using VEZF1 model\n",
      "2021-07-19 15:18:13 | INFO | No match for PRDM4 based on motifs\n",
      "2021-07-19 15:18:13 | INFO | No related TF found for PRDM4, using general model\n",
      "2021-07-19 15:18:13 | INFO | No match for ATF6 based on motifs\n",
      "2021-07-19 15:18:13 | INFO | No related TF found for ATF6, using general model\n",
      "2021-07-19 15:18:13 | INFO | No match for ATF6B based on motifs\n",
      "2021-07-19 15:18:13 | INFO | No related TF found for ATF6B, using general model\n",
      "2021-07-19 15:18:14 | INFO | Using POU6F2 motif with PDX1 model weights\n",
      "2021-07-19 15:18:14 | INFO | No match for ZIM3 based on motifs\n",
      "2021-07-19 15:18:14 | INFO | No related TF found for ZIM3, using general model\n",
      "2021-07-19 15:18:15 | INFO | No match for ZNF320 based on motifs\n",
      "2021-07-19 15:18:15 | INFO | No related TF found for ZNF320, using general model\n",
      "2021-07-19 15:18:15 | INFO | Using NFE2L1 model\n",
      "2021-07-19 15:18:15 | INFO | No match for ZBTB4 based on motifs\n",
      "2021-07-19 15:18:15 | INFO | No related TF found for ZBTB4, using general model\n",
      "2021-07-19 15:18:16 | INFO | No match for ZNF431 based on motifs\n",
      "2021-07-19 15:18:16 | INFO | No related TF found for ZNF431, using general model\n",
      "2021-07-19 15:18:16 | INFO | No match for PURA based on motifs\n",
      "2021-07-19 15:18:16 | INFO | No related TF found for PURA, using general model\n",
      "2021-07-19 15:18:16 | INFO | No match for ZFP57 based on motifs\n",
      "2021-07-19 15:18:16 | INFO | No related TF found for ZFP57, using general model\n",
      "2021-07-19 15:18:17 | INFO | No match for ZNF136 based on motifs\n",
      "2021-07-19 15:18:17 | INFO | No related TF found for ZNF136, using general model\n",
      "2021-07-19 15:18:17 | INFO | No match for ZNF134 based on motifs\n",
      "2021-07-19 15:18:17 | INFO | No related TF found for ZNF134, using general model\n",
      "2021-07-19 15:18:18 | INFO | Using TFAP2D motif with TFAP2C model weights\n",
      "2021-07-19 15:18:18 | INFO | No match for PRDM5 based on motifs\n",
      "2021-07-19 15:18:18 | INFO | No related TF found for PRDM5, using general model\n",
      "2021-07-19 15:18:18 | INFO | No match for ARNT2 based on motifs\n",
      "2021-07-19 15:18:18 | INFO | No related TF found for ARNT2, using general model\n",
      "2021-07-19 15:18:19 | INFO | No match for ZBTB49 based on motifs\n",
      "2021-07-19 15:18:19 | INFO | No related TF found for ZBTB49, using general model\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-07-19 15:18:19 | INFO | Using ZNF652 model\n",
      "2021-07-19 15:18:20 | INFO | Using ZBTB47 motif with ZNF652 model weights\n",
      "2021-07-19 15:18:20 | INFO | No match for ZSCAN22 based on motifs\n",
      "2021-07-19 15:18:20 | INFO | No related TF found for ZSCAN22, using general model\n",
      "2021-07-19 15:18:20 | INFO | No match for NR2E3 based on motifs\n",
      "2021-07-19 15:18:20 | INFO | No related TF found for NR2E3, using general model\n",
      "2021-07-19 15:18:21 | INFO | No match for GTF2I based on motifs\n",
      "2021-07-19 15:18:21 | INFO | No related TF found for GTF2I, using general model\n",
      "2021-07-19 15:18:21 | INFO | Using ZBTB1 model\n",
      "2021-07-19 15:18:21 | INFO | No match for PATZ1 based on motifs\n",
      "2021-07-19 15:18:21 | INFO | No related TF found for PATZ1, using general model\n",
      "2021-07-19 15:18:22 | INFO | No match for SP100 based on motifs\n",
      "2021-07-19 15:18:22 | INFO | No related TF found for SP100, using general model\n",
      "2021-07-19 15:18:22 | INFO | No match for ZNF582 based on motifs\n",
      "2021-07-19 15:18:22 | INFO | No related TF found for ZNF582, using general model\n",
      "2021-07-19 15:18:23 | INFO | No match for ZNF41 based on motifs\n",
      "2021-07-19 15:18:23 | INFO | No related TF found for ZNF41, using general model\n",
      "2021-07-19 15:18:23 | INFO | Using ZNF24 model\n",
      "2021-07-19 15:18:23 | INFO | No match for NR6A1 based on motifs\n",
      "2021-07-19 15:18:23 | INFO | No related TF found for NR6A1, using general model\n",
      "2021-07-19 15:18:24 | INFO | No match for INSM1 based on motifs\n",
      "2021-07-19 15:18:24 | INFO | No related TF found for INSM1, using general model\n",
      "2021-07-19 15:18:24 | INFO | Using SCX motif with NEUROD1 model weights\n",
      "2021-07-19 15:18:25 | INFO | No match for CREBZF based on motifs\n",
      "2021-07-19 15:18:25 | INFO | No related TF found for CREBZF, using general model\n",
      "2021-07-19 15:18:25 | INFO | Using RHOXF2 motif with MIXL1 model weights\n",
      "2021-07-19 15:18:25 | INFO | No match for FOXH1 based on motifs\n",
      "2021-07-19 15:18:25 | INFO | No related TF found for FOXH1, using general model\n",
      "2021-07-19 15:18:26 | INFO | No match for ZNF816 based on motifs\n",
      "2021-07-19 15:18:26 | INFO | No related TF found for ZNF816, using general model\n",
      "2021-07-19 15:18:26 | INFO | No match for ZFP3 based on motifs\n",
      "2021-07-19 15:18:26 | INFO | No related TF found for ZFP3, using general model\n",
      "2021-07-19 15:18:26 | INFO | Using ZNF18 model\n",
      "2021-07-19 15:18:27 | INFO | No match for HDX based on motifs\n",
      "2021-07-19 15:18:27 | INFO | No related TF found for HDX, using general model\n",
      "2021-07-19 15:18:27 | INFO | No match for PROX1 based on motifs\n",
      "2021-07-19 15:18:27 | INFO | No related TF found for PROX1, using general model\n",
      "2021-07-19 15:18:28 | INFO | No match for ZNF329 based on motifs\n",
      "2021-07-19 15:18:28 | INFO | No related TF found for ZNF329, using general model\n",
      "2021-07-19 15:18:28 | INFO | No match for ZNF770 based on motifs\n",
      "2021-07-19 15:18:28 | INFO | No related TF found for ZNF770, using general model\n",
      "2021-07-19 15:18:28 | INFO | No match for ZSCAN31 based on motifs\n",
      "2021-07-19 15:18:28 | INFO | No related TF found for ZSCAN31, using general model\n",
      "2021-07-19 15:18:29 | INFO | No match for ZSCAN26 based on motifs\n",
      "2021-07-19 15:18:29 | INFO | No related TF found for ZSCAN26, using general model\n",
      "2021-07-19 15:18:29 | INFO | Using ZNF282 model\n",
      "2021-07-19 15:18:30 | INFO | No match for ZNF768 based on motifs\n",
      "2021-07-19 15:18:30 | INFO | No related TF found for ZNF768, using general model\n",
      "2021-07-19 15:18:30 | INFO | No match for ZNF317 based on motifs\n",
      "2021-07-19 15:18:30 | INFO | No related TF found for ZNF317, using general model\n",
      "2021-07-19 15:18:30 | INFO | No match for NPAS4 based on motifs\n",
      "2021-07-19 15:18:30 | INFO | No related TF found for NPAS4, using general model\n",
      "2021-07-19 15:18:31 | INFO | No match for ZNF708 based on motifs\n",
      "2021-07-19 15:18:31 | INFO | No related TF found for ZNF708, using general model\n",
      "2021-07-19 15:18:31 | INFO | No match for ZNF547 based on motifs\n",
      "2021-07-19 15:18:31 | INFO | No related TF found for ZNF547, using general model\n",
      "2021-07-19 15:18:32 | INFO | No match for ZNF324 based on motifs\n",
      "2021-07-19 15:18:32 | INFO | No related TF found for ZNF324, using general model\n",
      "2021-07-19 15:18:32 | INFO | No match for ZNF121 based on motifs\n",
      "2021-07-19 15:18:32 | INFO | No related TF found for ZNF121, using general model\n",
      "2021-07-19 15:18:32 | INFO | No match for ZNF250 based on motifs\n",
      "2021-07-19 15:18:32 | INFO | No related TF found for ZNF250, using general model\n",
      "2021-07-19 15:18:33 | INFO | Using IKZF3 motif with IKZF1 model weights\n",
      "2021-07-19 15:18:33 | INFO | No match for ZNF502 based on motifs\n",
      "2021-07-19 15:18:33 | INFO | No related TF found for ZNF502, using general model\n",
      "2021-07-19 15:18:33 | INFO | No match for ZNF165 based on motifs\n",
      "2021-07-19 15:18:33 | INFO | No related TF found for ZNF165, using general model\n",
      "2021-07-19 15:18:34 | INFO | No match for ZNF214 based on motifs\n",
      "2021-07-19 15:18:34 | INFO | No related TF found for ZNF214, using general model\n",
      "2021-07-19 15:18:34 | INFO | No match for ZBTB48 based on motifs\n",
      "2021-07-19 15:18:34 | INFO | No related TF found for ZBTB48, using general model\n",
      "2021-07-19 15:18:35 | INFO | Using ZKSCAN1 model\n",
      "2021-07-19 15:18:35 | INFO | No match for ZNF331 based on motifs\n",
      "2021-07-19 15:18:35 | INFO | No related TF found for ZNF331, using general model\n",
      "2021-07-19 15:18:35 | INFO | No match for ZFP28 based on motifs\n",
      "2021-07-19 15:18:35 | INFO | No related TF found for ZFP28, using general model\n",
      "2021-07-19 15:18:36 | INFO | No match for ZNF449 based on motifs\n",
      "2021-07-19 15:18:36 | INFO | No related TF found for ZNF449, using general model\n",
      "2021-07-19 15:18:36 | INFO | No match for ZNF341 based on motifs\n",
      "2021-07-19 15:18:36 | INFO | No related TF found for ZNF341, using general model\n",
      "2021-07-19 15:18:36 | INFO | No match for ZNF490 based on motifs\n",
      "2021-07-19 15:18:36 | INFO | No related TF found for ZNF490, using general model\n",
      "2021-07-19 15:18:37 | INFO | No match for ZNF784 based on motifs\n",
      "2021-07-19 15:18:37 | INFO | No related TF found for ZNF784, using general model\n",
      "2021-07-19 15:18:37 | INFO | No match for ZNF680 based on motifs\n",
      "2021-07-19 15:18:37 | INFO | No related TF found for ZNF680, using general model\n",
      "2021-07-19 15:18:38 | INFO | No match for MYT1L based on motifs\n",
      "2021-07-19 15:18:38 | INFO | No related TF found for MYT1L, using general model\n",
      "2021-07-19 15:18:38 | INFO | No match for ZNF257 based on motifs\n",
      "2021-07-19 15:18:38 | INFO | No related TF found for ZNF257, using general model\n",
      "2021-07-19 15:18:38 | INFO | No match for ZNF628 based on motifs\n",
      "2021-07-19 15:18:38 | INFO | No related TF found for ZNF628, using general model\n",
      "2021-07-19 15:18:39 | INFO | No match for ZNF418 based on motifs\n",
      "2021-07-19 15:18:39 | INFO | No related TF found for ZNF418, using general model\n",
      "2021-07-19 15:18:39 | INFO | No match for ZNF140 based on motifs\n",
      "2021-07-19 15:18:39 | INFO | No related TF found for ZNF140, using general model\n",
      "2021-07-19 15:18:40 | INFO | Using ENO1 motif with MYC model weights\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSSCs\n",
      "LNPCs\n",
      "LPCs\n",
      "StC\n",
      "MEC\n"
     ]
    }
   ],
   "source": [
    "# lets do ananse binding for your cell populations\n",
    "i = 0\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}\").mkdir(parents=True, exist_ok=True)       \n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/binding.h5\"): \n",
    "        print(f'running ananse binding for celltype {cell_id} using the scATAC datatypes with the ATAC merged peaks of all joined peaks')\n",
    "        #with open(f'{output_dir}/{cell_id}/myfile.txt', 'w') as fp: \n",
    "       #     pass\n",
    "        sp.check_call(f'ananse binding '\n",
    "                f'-g {genome_path} '\n",
    "                f'-A /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/merge_peaks/bamfiles/{cell_id}.bam '\n",
    "                f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/merge_peaks/accesible_summits.bed '\n",
    "                f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/{cell_id} '\n",
    "                f'--pfmscorefile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outsscan_results_acc.scanfile '\n",
    "                f'-n 4',shell = True)     \n",
    "\n",
    "#f'2> {output_dir}/{cell_id}/ananse_binding_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generating the binding.h5 file for ESC data ananse 0.3.0\n",
    "Path(f\"{output_dir}/ESC\").mkdir(parents=True, exist_ok=True)       \n",
    "if not os.path.exists(f\"{output_dir}/ESC/binding.h5\"): \n",
    "    sp.check_call(f'ananse binding '\n",
    "        f'-g {genome_path} '\n",
    "        f'-A /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/H3K27ac/results/final_bam/hg38-GSM466732.samtools-coordinate.bam '\n",
    "        f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/all_peaks.tsv '\n",
    "        f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC '\n",
    "        f'--pfmscorefile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3scan_results_ESC.scanfile '\n",
    "        f'-n 4 '   \n",
    "        f'2> {output_dir}/ESC/ananse_binding_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse network\n",
    "\n",
    "Generate GRN's based on the binding.tsv output and RNAseq expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make one TPM file per sample using the big count matrix containing all samples; already done\n",
    "# ensembl2hugo =  pd.read_table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/genomepy_genomes/GRCh38.p13/gene_id2name.tsv',sep = '\\t', comment ='_',names = ['ensembl','HUGO'])\n",
    "# for index, cell_type in sample_data.iterrows():\n",
    "#     for binding_mode in ['only_H3K27ac','Remap_H3K27ac']:\n",
    "#         cell_id = sample_data.iloc[index,0]\n",
    "#         TPM_matrix_file = sample_data.TPM_matrix[index]\n",
    "#         TPM_matrix = pd.read_table(TPM_matrix_file,sep = '\\t',index_col = 0)\n",
    "        \n",
    "#         #convert matrix to gene name; not nessesary if you do it beforehand\n",
    "#         #hugo_tpm_matrix = TPM_matrix.merge(ensembl2hugo, left_index = True, right_on = 'ensembl', how = 'inner')\n",
    "#         #hugo_tpm_matrix.index = hugo_tpm_matrix.HUGO\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop_duplicates()\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop_duplicates(subset='HUGO', keep=\"first\")\n",
    "#         #hugo_tpm_matrix = hugo_tpm_matrix.drop(['ensembl','HUGO'], axis = 1)\n",
    "\n",
    "#         TPM_files = sample_data.TPM_names[index].split(' ')\n",
    "#         Path(f\"{output_dir}/{binding_mode}/{cell_id}/TPM_files/\").mkdir(parents=True, exist_ok=True) \n",
    "#         for sample in TPM_files:\n",
    "#             sample_matrix = hugo_tpm_matrix[sample]\n",
    "#             sample_matrix = pd.DataFrame(sample_matrix)\n",
    "#             sample_matrix[['target_id']] = sample_matrix.index\n",
    "#             sample_matrix[['tpm']] = sample_matrix[[sample]]\n",
    "#             sample_matrix[['target_id','tpm']]\n",
    "#             #save sample TPM file\n",
    "#             sample_matrix[['target_id','tpm']].to_csv(f\"{output_dir}/{binding_mode}/{cell_id}/TPM_files/FPM_{sample}.tsv\", sep = '\\t', index = False, header = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>control_t0B</th>\n",
       "      <th>control_t0A</th>\n",
       "      <th>tcdd_t2A</th>\n",
       "      <th>tcdd_t2B</th>\n",
       "      <th>control_t24A</th>\n",
       "      <th>control_t24B</th>\n",
       "      <th>tcdd_t24A</th>\n",
       "      <th>tcdd_t24B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUGO</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>17.149720</td>\n",
       "      <td>10.471415</td>\n",
       "      <td>10.595141</td>\n",
       "      <td>7.870116</td>\n",
       "      <td>6.327820</td>\n",
       "      <td>7.063620</td>\n",
       "      <td>7.979823</td>\n",
       "      <td>6.325595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>142.842093</td>\n",
       "      <td>107.685809</td>\n",
       "      <td>69.834197</td>\n",
       "      <td>68.042334</td>\n",
       "      <td>55.006878</td>\n",
       "      <td>53.782104</td>\n",
       "      <td>49.323561</td>\n",
       "      <td>41.976615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>4.579877</td>\n",
       "      <td>3.434859</td>\n",
       "      <td>2.408801</td>\n",
       "      <td>1.757298</td>\n",
       "      <td>2.696829</td>\n",
       "      <td>2.355813</td>\n",
       "      <td>2.560131</td>\n",
       "      <td>2.280325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>24.530547</td>\n",
       "      <td>17.480404</td>\n",
       "      <td>16.068732</td>\n",
       "      <td>11.890818</td>\n",
       "      <td>1.880808</td>\n",
       "      <td>1.570157</td>\n",
       "      <td>6.456277</td>\n",
       "      <td>6.035803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC064824.1</th>\n",
       "      <td>0.056404</td>\n",
       "      <td>0.107946</td>\n",
       "      <td>0.085177</td>\n",
       "      <td>0.021252</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016335</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.032467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL136225.2</th>\n",
       "      <td>0.056168</td>\n",
       "      <td>0.135234</td>\n",
       "      <td>0.090534</td>\n",
       "      <td>0.037190</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115257</td>\n",
       "      <td>0.082025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC004636.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC007687.1</th>\n",
       "      <td>0.355343</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.238526</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.237341</td>\n",
       "      <td>0.425962</td>\n",
       "      <td>0.032896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL353795.4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.571044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>59143 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            control_t0B  control_t0A   tcdd_t2A   tcdd_t2B  control_t24A  \\\n",
       "HUGO                                                                       \n",
       "TSPAN6        17.149720    10.471415  10.595141   7.870116      6.327820   \n",
       "TNMD           0.000000     0.000000   0.000000   0.000000      0.000000   \n",
       "DPM1         142.842093   107.685809  69.834197  68.042334     55.006878   \n",
       "SCYL3          4.579877     3.434859   2.408801   1.757298      2.696829   \n",
       "C1orf112      24.530547    17.480404  16.068732  11.890818      1.880808   \n",
       "...                 ...          ...        ...        ...           ...   \n",
       "AC064824.1     0.056404     0.107946   0.085177   0.021252      0.000000   \n",
       "AL136225.2     0.056168     0.135234   0.090534   0.037190      0.000000   \n",
       "AC004636.1     0.000000     0.000000   0.000000   0.000000      0.000000   \n",
       "AC007687.1     0.355343     0.000000   0.000000   0.238526      0.000000   \n",
       "AL353795.4     0.000000     0.000000   0.000000   0.571044      0.000000   \n",
       "\n",
       "            control_t24B  tcdd_t24A  tcdd_t24B  \n",
       "HUGO                                            \n",
       "TSPAN6          7.063620   7.979823   6.325595  \n",
       "TNMD            0.000000   0.000000   0.000000  \n",
       "DPM1           53.782104  49.323561  41.976615  \n",
       "SCYL3           2.355813   2.560131   2.280325  \n",
       "C1orf112        1.570157   6.456277   6.035803  \n",
       "...                  ...        ...        ...  \n",
       "AC064824.1      0.016335   0.000000   0.032467  \n",
       "AL136225.2      0.000000   0.115257   0.082025  \n",
       "AC004636.1      0.000000   0.000000   0.000000  \n",
       "AC007687.1      0.237341   0.425962   0.032896  \n",
       "AL353795.4      0.000000   0.000000   0.000000  \n",
       "\n",
       "[59143 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#hugo_tpm_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CBtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CjS\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CjStpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSB\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSB/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSBtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSSCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CSSCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/CSSCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LNPCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LNPCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LNPCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LPCs\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/LPCs/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/LPCstpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "StC\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/StC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/StCtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "MEC\n",
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/MEC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/jupyter_notebooks/MECtpm.tsv\n",
      "and the output file full_network_includeprom.txt\n"
     ]
    }
   ],
   "source": [
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    binding_file = f\"{output_dir}/{cell_id}/binding.h5\"\n",
    "    print(f'running ananse binding using the enhancer file {binding_file}')\n",
    "    file_name = sample_data.iloc[index,4]\n",
    "    print(f'together with the TPM file {file_name}')\n",
    "    print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"): \n",
    "        sp.check_call(f'nice -15 ananse network '\n",
    "            f'-e {file_name} '\n",
    "            f'-b {binding_file} '\n",
    "            f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/{cell_id}/full_network_includeprom.txt '\n",
    "            f'-g hg38 '\n",
    "            f'-n 1',shell = True)\n",
    "\n",
    "#f'2> {output_dir}/{cell_id}/ananse_network_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/binding.h5\n",
      "together with the TPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/tpm_ESC.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "[########################################] | 100% Completed | 52.2s\r"
     ]
    }
   ],
   "source": [
    "# generate the ESC network file from the binding.h5 file\n",
    "binding_file = f\"{output_dir}/ESC/binding.h5\"\n",
    "print(f'running ananse binding using the enhancer file {binding_file}')\n",
    "file_name = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/tpm_ESC.tsv\"\n",
    "print(f'together with the TPM file {file_name}')\n",
    "print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "if not os.path.exists(f\"{output_dir}/ESC/full_network_includeprom.txt\"): \n",
    "    sp.check_call(f'nice -15 ananse network '\n",
    "        f'-e {file_name} '\n",
    "        f'-b {binding_file} '\n",
    "        f'-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt '\n",
    "        f'-g hg38 '\n",
    "        f'-n 1 '\n",
    "        f'2> {output_dir}/ESC/ananse_network_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse Influence\n",
    "\n",
    "Lets finally run Ananse influence to predict key TFs that differ between cell types.\n",
    "it will compare each network with the networks listed in the 'comparison' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[43334, 43335, 43334, 43334]\n",
      "[42583, 42583, 42583, 42583]\n"
     ]
    }
   ],
   "source": [
    "# check threading and os delay\n",
    "import joblib\n",
    "from joblib import Parallel, delayed\n",
    "from joblib import parallel_backend\n",
    "\n",
    "print(Parallel(n_jobs=2)(delayed(os.getpid)() for i in range(4)))\n",
    "\n",
    "with parallel_backend('threading'):\n",
    "    print(Parallel(n_jobs=2)(delayed(os.getpid)() for i in range(4)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "100000\n"
     ]
    }
   ],
   "source": [
    "# adapting the recursion limit\n",
    "print(sys.getrecursionlimit())\n",
    "sys.setrecursionlimit(100000)\n",
    "print(sys.getrecursionlimit())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CB/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20210720/CBESCpseudobulkpadj.tsv and the output file influence.txt\n",
      "CjS\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20210720/CjSESCpseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'nice -15 ananse influence -s /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt -t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/full_network_includeprom.txt -d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20210720/CjSESCpseudobulkpadj.tsv -i 100000 -o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/ESCCjS_influence_100000_022/influence.txt -n 1 2> /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/ananse_influence_log_newESC_022.txt' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m/scratch/julian/tmp/anYH/ipykernel_20554/1909431129.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{output_dir}/{cell_id}/ESC{cell_id}_influence_100000_022/influence.txt\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m         sp.check_call(f'nice -15 ananse influence '\n\u001b[0m\u001b[1;32m     14\u001b[0m             \u001b[0;34mf'-s {net2} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m             \u001b[0;34mf'-t {network_file} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse030/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    371\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcmd\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    372\u001b[0m             \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpopenargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 373\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    374\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    375\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'nice -15 ananse influence -s /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt -t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/full_network_includeprom.txt -d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/R/20210720/CjSESCpseudobulkpadj.tsv -i 100000 -o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/ESCCjS_influence_100000_022/influence.txt -n 1 2> /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/CjS/ananse_influence_log_newESC_022.txt' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs3/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/ESC{cell_id}_influence_100000_022\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC{cell_id}_influence_100000_022/influence.txt\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-s {net2} '\n",
    "            f'-t {network_file} '\n",
    "            f'-d {file_name} '\n",
    "            f'-i 100000 '\n",
    "            #f'--plot '    \n",
    "            f'-o {output_dir}/{cell_id}/ESC{cell_id}_influence_100000_022/influence.txt '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_newESC_022.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comparing KC_0h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_24h_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_24h_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD30_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD30_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_2h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_2h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_0and24_CTR_to_KC_TCDD90_24h_RNA\n",
      "running Deseq2\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "running influence for KC_0and24_CTR_to_KC_TCDD90_24h_RNA with the datatype of only_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_0h_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_24h_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_24h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_24h_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "comparing KC_TCDD30_2h_RNA_to_KC_0and24_CTR\n",
      "running Deseq2\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0and24_CTR with the datatype of Remap_H3K27ac and include-promoter\n",
      "running influence for KC_TCDD30_2h_RNA_to_KC_0and24_CTR with the datatype of Remap_H3K27ac and include-promoter\n"
     ]
    }
   ],
   "source": [
    "### delete in full??\n",
    "\n",
    "ensembl2hugo =  pd.read_table('/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/genomepy_genomes/GRCh38.p13/gene_id2name.tsv',sep = '\\t', comment ='_',names = ['ensembl','HUGO'])\n",
    "\n",
    "\n",
    "\n",
    "nice -15 ananse influence \n",
    "-s /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/peakpred_V4/only_ATAC/ESCs/full_network_include-promoter.txt \n",
    "-t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs/CjS/full_network_includeprom.txt \n",
    "-d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/pseudobulk_rnaseq/CjSESC_degenes.tsv \n",
    "-o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/lako2021/ANANSE/outs/CjS/ESCCjS_influence/influence.txt -n 10\n",
    "\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    for binding_mode in ['Remap_H3K27ac','only_H3K27ac']:\n",
    "        comparisons = sample_data.compare_with[index]\n",
    "        cell_id = sample_data.iloc[index,0]\n",
    "        #get the count data files for the source cell type\n",
    "        count_files_source_cell_type = sample_data.TPM_names[index].split(' ')\n",
    "        if not comparisons != comparisons: #do stuff if there is a string in the comparisons column\n",
    "            for comparison in comparisons.split(' '):\n",
    "                #get the count data files for the target cell type\n",
    "                target_data = sample_data.loc[sample_data['cell_type'] == comparison]\n",
    "                target_data = target_data.reset_index(drop=True)\n",
    "                count_files_target_cell_type = target_data.TPM_names[0].split(' ')\n",
    "                #lets run deseq between the sample comparisons \n",
    "                comp_name = f'{cell_id}_to_{comparison}'\n",
    "                print(f'comparing {comp_name}')\n",
    "                if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}\"):\n",
    "                    Path(f\"{output_dir}/{binding_mode}/influence/{comp_name}\").mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "                #lets load the countmatrix, and remove all columns that don't contain one of the samples\n",
    "                count_matrix = pd.read_table(sample_data.count_table_files[index])\n",
    "                #count_matrix[(count_files_source_cell_type + count_files_target_cell_type)]\n",
    "                #switch ensembl to HUGO gene names\n",
    "                hugo_count_matrix = count_matrix.merge(ensembl2hugo, left_on = 'gene', right_on = 'ensembl', how = 'inner')\n",
    "                hugo_count_matrix.index = hugo_count_matrix.HUGO\n",
    "                hugo_count_matrix = hugo_count_matrix.drop_duplicates()\n",
    "                hugo_count_matrix = hugo_count_matrix.drop_duplicates(subset='HUGO', keep=\"first\")\n",
    "                hugo_count_matrix = hugo_count_matrix.drop(['ensembl','HUGO','gene'], axis = 1)\n",
    "                #hugo_count_matrix.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv\", sep = '\\t', index = True, header = True)\n",
    "                hugo_count_matrix = hugo_count_matrix[(count_files_source_cell_type + count_files_target_cell_type)]\n",
    "                hugo_count_matrix.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv\", sep = '\\t', index = True, header = True)\n",
    "                #make a sample file for deseq2 \n",
    "                sample_names = (count_files_source_cell_type + count_files_target_cell_type)\n",
    "                annotation = [\"source\"] * len(count_files_source_cell_type) + [\"target\"] * len(count_files_target_cell_type)\n",
    "                data= {'samples':sample_names, 'condition':annotation}\n",
    "                sample_info_df = pd.DataFrame(data)\n",
    "                sample_info_df.to_csv(f\"{output_dir}/{binding_mode}/influence/{comp_name}/sample_info.tsv\", sep = ',', index = False, header = True)\n",
    "                #run Deseq2 \n",
    "                #if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv\"):\n",
    "                \n",
    "                if not os.path.exists(f\"{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv\"):\n",
    "                    print('running Deseq2')\n",
    "                    sp.check_call(\n",
    "                        f'nice -5 Rscript --vanilla '\n",
    "                        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/analysis/RunDeseq2.R '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/count_matrix.tsv '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/sample_info.tsv '\n",
    "                        f'{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv '\n",
    "                        f'2> {output_dir}/{binding_mode}/influence/{comp_name}/deseq2log.txt',\n",
    "                        shell = True)\n",
    "\n",
    "                    DEG_table = pd.read_table(f'{output_dir}/{binding_mode}/influence/{comp_name}/DEGs.csv', \n",
    "                            sep = ',', comment = '#')\n",
    "                    DEG_table['resid'] = DEG_table['region']\n",
    "                    DEG_table = DEG_table.dropna()\n",
    "                    DEG_table[['resid','log2FoldChange','padj']].to_csv(f'{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv', sep = '\\t', index = False, header = True)\n",
    "                    DEG_file = f'{output_dir}/{binding_mode}/influence/{comp_name}/DEG.tsv'\n",
    "\n",
    "                for weight_modus in ['include-promoter']:\n",
    "                    for n_edges in [100000, 500000]:\n",
    "                        source_network = f'{output_dir}/{binding_mode}/{cell_id}/full_network_{weight_modus}.txt'\n",
    "                        target_network = f'{output_dir}/{binding_mode}/{comparison}/full_network_{weight_modus}.txt'\n",
    "                        influence_output_name = f'{output_dir}/{binding_mode}/influence/{comp_name}/{weight_modus}/{n_edges}/'\n",
    "                        if not os.path.exists(f\"{influence_output_name}\"):\n",
    "                            Path(f\"{influence_output_name}\").mkdir(parents=True, exist_ok=True)\n",
    "                        if not os.path.exists(f\"{influence_output_name}influence.txt\"):\n",
    "                            print(f'running influence for {comp_name} with the datatype of {binding_mode} and {weight_modus}')\n",
    "                            sp.check_call(f'nice -15 ananse influence '\n",
    "                                f'-s {source_network} '\n",
    "                                f'-t {target_network} '\n",
    "                                f'-d {DEG_file} '\n",
    "                                f'-p '\n",
    "                                f'-o {influence_output_name}influence.txt ' \n",
    "                                f'--edges {n_edges} '\n",
    "                                f'-n 12 '\n",
    "                                f'2> {influence_output_name}log.txt',shell = True)  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
